---
title: "GWAS results annotation & enrichment"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(glue)
library(magrittr)
library(stringr)
library(tidylog)
library(GALLO)
library(gprofiler2)

source(here::here("source_functions/gcif.R"))

options(scipen = 0)

# I can't remember numbers and have to re-calculate this every time so
q_fdr5 <- -log10(.05)
q_fdr1 <- -log10(.01)
```

# Notes & questions

* SNP1101 `freq` column is allele frequency not genotype frequency: $\frac{g_2+\frac{1}{2}g_1}{g_0+g_1+g_2}$ or $\frac{2g_{2}+g_{1}}{2(g_0+g_1+g_2)}$

---

# Setup

## Upstream results

* SNP1101

```{r, warning = FALSE, message=FALSE}
fixed9_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(adj_p = gcif(df = .,
                      adjust_p = TRUE,
                      p_col = p_value),
         q = qvalue::qvalue(adj_p)$qvalues,
         neglog10q = -log10(q))
  
```

* COJO

```{r}
years_jma <- 
  read_table2(here::here("data/derived_data/cojo/years/years_subset_adj.jma.cojo")) %>% 
  janitor::clean_names() %>% 
  rename(pos = bp)
```

* Metasoft GxE

```{r, warning=FALSE, message=FALSE}
gxe_metasoft_dl <- 
  read_table2(here::here("data/derived_data/metasoft/day_length/metasoft_out.day_length_adj.txt"),
              col_types = cols(.default = "d", RSID = "c")) %>% 
  janitor::clean_names() %>% 
  select(-pvalue_be, -contains("fe")) %>%
  left_join(read_table2(here::here("data/derived_data/metasoft/day_length/metasoft_out.day_length_adj.txt"),
                        skip = 1,
                        col_types = "c---------------dddddddd", 
                        col_names = c("rsid", "p16", "p17", "p18", "p19", "m16", "m17", "m18", "m19"))) %>% 
  mutate(chr = as.numeric(str_extract(rsid, "[[:digit:]]{1,2}(?=:)")),
         pos = as.numeric(str_extract(rsid, "(?<=[[:digit:]]{1,2}:)[[:digit:]]+")),
         neglog10q_re2 = -log10(qvalue::qvalue(pvalue_re2)$qvalues))
```

## GFF and GTF files

```{r}
gtf <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.101.gtf"),
                 file_type = "gtf")
```

```{r}
gff <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.QTL.gff"),
                 file_type = "gff")
```

# Annotation

## Genes

```{r, eval = FALSE}
gene0 <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = fixed9_snp1101 %>% 
                                   filter(neglog10q >= q_fdr5) %>% 
                                   select(CHR = chr, BP = pos),
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 0) %>% 
  mutate(which = "fixed9") %>% 
  bind_rows(find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = years_jma %>% 
                                             select(CHR = chr, BP = pos),
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 0) %>% 
              mutate(which = "cojo")) %>% 
  bind_rows(find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = gxe_metasoft_dl %>% 
                                             filter(neglog10q_re2 >= q_fdr5) %>% 
                                             select(CHR = chr, BP = pos),
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 0) %>% 
              mutate(which = "gxe"))
```

```{r}
gene10k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = fixed9_snp1101 %>% 
                                   filter(neglog10q >= q_fdr5) %>% 
                                   select(CHR = chr, BP = pos),
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 10000) %>% 
  mutate(which = "fixed9") %>% 
  bind_rows(find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = years_jma %>% 
                                             select(CHR = chr, BP = pos),
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 10000) %>% 
              mutate(which = "cojo")) %>% 
  bind_rows(find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = gxe_metasoft_dl %>% 
                                             filter(neglog10q_re2 >= q_fdr5) %>% 
                                             select(CHR = chr, BP = pos),
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 10000) %>% 
              mutate(which = "gxe"))
```

## QTLs

```{r}
qtl10k <-
  find_genes_qtls_around_markers(db_file = gff,
                                 marker_file = fixed9_snp1101 %>% 
                                   filter(neglog10q >= q_fdr5) %>% 
                                   select(CHR = chr, BP = pos),
                                 method = c("qtl"),
                                 marker = c("snp"),
                                 interval = 10000) %>% 
  mutate(which = "fixed9") %>% 
  bind_rows(find_genes_qtls_around_markers(db_file = gff,
                                           marker_file = gxe_metasoft_dl %>% 
                                             filter(neglog10q_re2 >= q_fdr5) %>% 
                                             select(CHR = chr, BP = pos),
                                           method = c("qtl"),
                                           marker = c("snp"),
                                           interval = 10000) %>% 
              mutate(which = "gxe"))
```

# Calculate enrichments

## Genes

```{r}
gene_enrich10k <-
  c("fixed9", "gxe") %>% 
  purrr::set_names() %>% 
  purrr::map(~ gene10k %>%
               filter(which == .x) %>% 
               distinct(gene_id) %>% 
               pull(gene_id) %>% 
               gost(organism = "btaurus",
                    significant = TRUE,
                    ordered = FALSE))
```

```{r}
gene_result10k <-
  c("fixed9", "gxe") %>% 
  map_dfr(~ gene_enrich10k[[.x]]$result %>% 
            mutate(which = .x))
```

## QTL

```{r}
qtl_enrich10k <-
  c("fixed9", "gxe") %>% 
  purrr::map_dfr(~ qtl_enrich(qtl_db = gff,
                              qtl_file = qtl10k %>% 
                                filter(which == .x),
                              qtl_type = "Name",
                              enrich_type = "genome",
                              padj = "BH") %>% 
                   mutate(which = .x))
```

# Gene ontology enrichment results

* `term_size` - number of genes that are annotated to the term
* `query_size` - number of genes that were included in the query. This might be different from the size of the original list if:
    + any genes were mapped to multiple Ensembl gene IDs
    + any genes failed to be mapped to Ensembl gene IDs
    + the parameter ordered_query = TRUE and the optimal cutoff for the term was found before the end of the query
    + the domain_scope was set to “annotated” or “custom”
* `intersection_size` - the number of genes in the input query that are annotated to the corresponding term
* `precision` - the proportion of genes in the input list that are annotated to the function (defined as intersection_size/query_size)

```{r}
listviewer::jsonedit(gene_enrich10k)
```

```{r}
gene_result10k %>% 
  select(which, term_name, p_value, everything(), -query, -significant) %>% 
  arrange(p_value)
```

# QTL enrichment results

```{r}
qtl_enrich10k %>%
  filter(0.05 >= adj.pval) %>% 
  arrange(which, adj.pval) %>% 
  select(which, adj.pval, everything())
```

## How many SNPs within each QTL?

```{r}
qtl_enrich10k %>% 
  filter(0.05 >= adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  arrange(adj.pval) %>% 
  left_join(qtl10k %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID)) %>% 
  group_by(QTL, N_QTLs, adj.pval, which) %>% 
  tidyr::nest() %>% 
  ungroup() %>% 
  mutate(unique_chr = purrr::map_chr(.x = data, 
                                 ~ .x %>% 
                                   arrange(CHR, BP) %>% 
                                   mutate(chr_bp = glue("{CHR}:{BP}")) %>% 
                                   distinct(chr_bp) %>% 
                                   pull(chr_bp) %>% 
                                   glue_collapse(sep = ", ")))
```

## White spotting

```{r}
qtl_enrich10k %>% 
  filter(which == "fixed9") %>% 
  filter(QTL == "White spotting") %>% 
  left_join(qtl10k %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID, which)) %>% 
  left_join(gene10k %>% 
              select(CHR, BP, gene_id, gene_name, which)) 
```

## Growth QTL

```{r}
qtl_enrich10k %>% 
  filter(0.05 >= adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  filter(QTL_type == "Production") %>% 
  left_join(qtl10k %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID, which)) %>% 
  left_join(gene10k %>% 
              select(CHR, BP, gene_id, gene_name, which)) 
```

```{r}
gene10k %>% 
  filter(str_detect(gene_name, "GH"))
```

# Positional candidates

```{r, eval = FALSE}
gene10k %>% 
  filter(which == "cojo")
```

## Chromosome 5

```{r, eval = FALSE}
years_jma %>% 
  filter(chr == 5) %>% 
  select(CHR = chr, BP = pos) %>% 
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = .,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 500000) %>% 
  mutate(ds = abs(BP-start_pos),
         de = abs(BP-end_pos),
         distance = if_else(de > ds, ds, de)) %>% 
  select(-ds, -de) %>% 
  arrange(distance) %>% 
  select(distance, everything())
```


```{r, eval = FALSE}
fixed9_snp1101 %>% 
  filter(neglog10q > q_fdr5) %>% 
  filter(chr == 5) %>% 
  filter(21000000 > pos) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(gene10k %>% 
              filter(which == "fixed9")) %>% 
  group_by(gene_name) %>% 
  tally(sort = TRUE)
```

### Lead SNP

```{r, eval = FALSE}
years_metasoft %>% 
  filter(chr == 5) %>% 
  # SNP(s) with largest neglog10q
  top_n(1, neglog10q) %>% 
  select(CHR = chr, BP = pos) %>% 
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = .,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 415000) %>% 
mutate(ds = abs(BP-start_pos),
         de = abs(BP-end_pos),
         distance = if_else(de > ds, ds, de)) %>% 
  select(-ds, -de) %>% 
  arrange(distance) %>% 
  select(distance, everything())
```


## Chromosome 23

```{r, eval=FALSE}
years_jma %>% 
  filter(chr == 23) %>% 
  select(CHR = chr, BP = pos) %>% 
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = .,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 100000) %>% 
  mutate(ds = abs(BP-start_pos),
         de = abs(BP-end_pos),
         distance = if_else(de > ds, ds, de)) %>% 
  select(-ds, -de) %>% 
  arrange(distance) %>% 
  select(distance, gene_name, everything())
```

```{r, eval = FALSE}
gene0k %>% 
  filter(CHR == 23) %>% 
  filter(gene_name == "PRL") %>% 
  filter(BP>= start_pos & end_pos >= BP)
```

```{r, eval = FALSE}
fixed9_snp1101 %>% 
  filter(chr == 23) %>%
  filter(pos > 30000000) %>% 
  filter(50000000 > pos) %>%
  left_join(full_gene10k %>% 
              select(gene_name, chr = CHR, pos = BP, start_pos, end_pos)) %>% 
  mutate(clr = case_when(pos >= start_pos & end_pos >= pos & gene_name == "CDKAL1" ~ "black",
                         TRUE ~ "#b9aa97")) %>% 
  ggplot(aes(x = pos,
             y = neglog10q,
             color = clr)) +
  geom_point(alpha = 0.75) +
  scale_color_identity() +
  scale_x_continuous(breaks = c(30000000, 35000000, 40000000, 45000000, 50000000),
                     labels = c("30 Mb", "35 Mb", "40 Mb", "45 Mb", "50 Mb")) +
  theme_classic() +
  theme(axis.title = element_text(size = 18),
        axis.title.y = element_text(margin = margin(t = 0,
                                                    r = 13,
                                                    b = 0,
                                                    l = 0)),
        axis.text = element_text(size = 14),
        legend.text = element_text(size = 14),
        axis.title.x = element_blank(),
        panel.background = element_rect(fill = "transparent",
                                        colour = NA),
        plot.background = element_rect(fill = "transparent",
                                       colour = NA)) +
  labs(x = NULL,
       y = latex2exp::TeX("$-log_{10}(q-value)$"),
       title = NULL) +
  geom_hline(yintercept = 1,
             color = "red",
             size = 0.5)
```

```{r, eval=FALSE}
ggsave(filename = here::here("figures/annotation/full_q_chr23.cdkal1.transparent.png"),
       width = 5,
       height = 3,
       bg = "transparent")
```

# Supplementary file

```{r, eval = FALSE}
supp1 <-
  all_gene10k %>% 
  select(-CHR) %>% 
  rename(snp_pos = BP, gene_start = start_pos, gene_end = end_pos) %>% 
  left_join(fixed9_snp1101 %>% 
              mutate(dataset = "full",
                     chr = as.character(chr)) %>% 
              filter(neglog10q > 1) %>% 
              select(dataset, snp_pos = pos, chr, snp_p = p_value) %>% 
              bind_rows(breeds_snp1101 %>% 
                          mutate(chr = as.character(chr)) %>% 
                          filter(neglog10p > 5) %>% 
                          select(dataset, snp_pos = pos, chr, snp_p = p_value))) %>% 
  mutate(within_gene = case_when(snp_pos >= gene_start & gene_end >= snp_pos ~ TRUE,
                                 TRUE ~ FALSE)) %>% 
  select(dataset, chr, snp_pos, snp_p, gene_start, gene_end, gene_id, gene_name, gene_biotype, within_gene)
```

# Genewalk

```{r, eval = FALSE}
gene50k %>% 
  filter(which == "gxe") %>% 
  mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>% 
  distinct(id) %>% 
  write_tsv(here::here("data/derived_data/string/string.gxe_50k_fdr1.txt"),
            col_names = FALSE)
```

```{r, eval = FALSE}
purrr::map(.x = c("fixed9", "cojo", "gxe"),
           ~ gene50k %>% 
             filter(which == .x) %>% 
             distinct(gene_id) %>% 
             pull(gene_id) %>% 
             gorth(query = .,
                   source_organism = "btaurus",
                   target_organism = "hsapiens",
                   mthreshold = 1,
                   filter_na = TRUE) %>% 
             select(ortholog_ensg) %>% 
             write_tsv(here::here(glue("data/derived_data/genewalk/{.x}_50k_fdr1/ortholog.{.x}_50k_fdr1.txt")),
                       col_names = FALSE))
```

```{r}
c("cojo", "fixed9", "gxe") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_csv(here::here(glue("data/derived_data/genewalk/{.x}_50k_fdr1/genewalk_results.csv"))),
                 .id = "which") %>% 
  filter_at(vars(contains("global")), any_vars(0.05 > .))
```

