---
title: "GWAS results annotation & enrichment"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(glue)
library(magrittr)
library(stringr)
library(tidylog)
library(GALLO)
library(gprofiler2)

options(scipen = 0)
```

# Notes & questions

* SNP1101 `freq` column is allele frequency not genotype frequency: $\frac{g_2+\frac{1}{2}g_1}{g_0+g_1+g_2}$ or $\frac{2g_{2}+g_{1}}{2(g_0+g_1+g_2)}$
* GCTA COJO requirements:
    + "`A1` needs to be the effect allele with `A2` being the other allele and `freq` should be the frequency of `A1`
    + `A1` and `A2` are A/T/C/G base calls
* COJO runs very slowly with `--cojo-gc` option + it calculated a very small inflation factor, took it out

---

# Setup

## SNP1101 raw GWAS results

```{r, warning = FALSE, message=FALSE}
fixed9_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(neglog10p = -log10(p_value),
         neglog10q = -log10(qvalue::qvalue(p_value)$qvalues),
         fdr_5 = qvalue::qvalue(p_value, fdr.level = 0.05)$significant, 
         fdr_1 = qvalue::qvalue(p_value, fdr.level = 0.01)$significant)
  
```

```{r}
q_fdr5 <-
  fixed9_snp1101 %>% 
  filter(fdr_5 == TRUE) %>% 
  arrange(neglog10q) %>% 
  slice(1) %>% 
  pull(neglog10q)
```

```{r}
q_fdr1 <-
  fixed9_snp1101 %>% 
  filter(fdr_1 == TRUE) %>% 
  arrange(neglog10q) %>% 
  slice(1) %>% 
  pull(neglog10q)
```

## Passing SNPs & COJO selected SNPs

```{r}
full_pass <-
  fixed9_snp1101 %>% 
  filter(fdr_1 == TRUE) %>% 
  select(CHR = chr, BP = pos)
```

```{r}
years_jma <- 
  read_table2(here::here("data/derived_data/cojo/years/years_subset.jma.cojo")) %>% 
  janitor::clean_names() %>% 
  rename(pos = bp)
```

```{r}
cojo_slct <- 
  years_jma %>% 
  select(CHR = chr, BP = pos)
```

## GFF and GTF files

```{r}
gtf <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.101.gtf"),
                 file_type = "gtf")
```

```{r}
gff <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.QTL.gff"),
                 file_type = "gff")
```

# Annotation

## Genes

### Passing SNPs

```{r}
full_gene50k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = full_pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 50000) 
```

```{r}
full_gene10k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = full_pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

```{r}
full_gene0 <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = full_pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 0) 
```

### COJO selected SNPs

```{r}
cojo_gene100k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = cojo_slct,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 100000) 
```

```{r}
cojo_gene50k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = cojo_slct,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 50000) 
```

```{r}
cojo_gene10k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = cojo_slct,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

```{r}
cojo_gene0 <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = cojo_slct,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 0) 
```

## QTLs

```{r}
full_qtl10k <-
  find_genes_qtls_around_markers(db_file = gff,
                                 marker_file = full_pass,
                                 method = c("qtl"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

# Calculate enrichments

## Genes

```{r}

full_gene_enrich10k <-
  full_gene10k %>% 
  distinct(gene_id) %>% 
  pull(gene_id) %>% 
  gost(organism = "btaurus",
       significant = TRUE)
```

```{r}

cojo_gene_enrich10k <-
  cojo_gene10k %>% 
  distinct(gene_id) %>% 
  pull(gene_id) %>% 
  gost(organism = "btaurus",
       significant = TRUE)
```

```{r}

cojo_gene_enrich50k <-
  cojo_gene50k %>% 
  distinct(gene_id) %>% 
  pull(gene_id) %>% 
  gost(organism = "btaurus",
       significant = TRUE)
```

## QTLs

```{r}

full_qtl_enrich10k <-
  qtl_enrich(qtl_db = gff,
             qtl_file = full_qtl10k,
             qtl_type = "Name",
             enrich_type = "genome",
             padj = "BH")
```

# Gene ontology enrichment results

* term_size - number of genes that are annotated to the term
* query_size - number of genes that were included in the query. This might be different from the size of the original list if:
    + any genes were mapped to multiple Ensembl gene IDs
    + any genes failed to be mapped to Ensembl gene IDs
    + the parameter ordered_query = TRUE and the optimal cutoff for the term was found before the end of the query
    + the domain_scope was set to “annotated” or “custom”
* intersection_size - the number of genes in the input query that are annotated to the corresponding term
* precision - the proportion of genes in the input list that are annotated to the function (defined as intersection_size/query_size)

```{r}
full_gene_enrich10k$result
```

# QTL enrichment results

```{r}
full_qtl_enrich10k %>%
  filter(0.05 > adj.pval) %>% 
  arrange(adj.pval) %>% 
  select(adj.pval, everything())
```

## How many SNPs within each QTL?

```{r, eval = FALSE}
full_qtl_enrich10k %>% 
  filter(0.05 >= adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  arrange(adj.pval) %>% 
  left_join(full_qtl10k %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID)) %>% 
  group_by(QTL, N_QTLs, adj.pval) %>% 
  summarise(n(), n_distinct(QTL_ID))
```

```{r,eval=FALSE}
full_qtl_enrich10k %>% 
  filter(0.05 >= adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  arrange(adj.pval) %>% 
  left_join(full_qtl10k %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID)) %>% 
  group_by(QTL, N_QTLs, adj.pval) %>% 
  tidyr::nest() %>% 
  ungroup() %>% 
  mutate(unique_chr = purrr::map_chr(.x = data, 
                                 ~ .x %>% 
                                   arrange(CHR, BP) %>% 
                                   mutate(chr_bp = glue("{CHR}:{BP}")) %>% 
                                   distinct(chr_bp) %>% 
                                   pull(chr_bp) %>% 
                                   glue_collapse(sep = ", ")))
```

## White spotting?

```{r, eval = FALSE}
full_qtl10k %>% 
  filter(trait_ID == "White spotting") %>% 
  select(-start_pos, -end_pos) %>% 
  left_join(full_gene0 %>% 
              filter(CHR == 22) %>% 
              select(CHR, BP, gene_name, gene_id, start_pos, end_pos)) %>% 
  filter(between(BP, start_pos, end_pos))
  
  
  
```

## Growth QTL

```{r, eval = FALSE}
full_qtl_enrich10k %>% 
  filter(0.05 >= adj.pval) %>% 
  filter(N_QTLs > 1) %>% 
  filter(QTL_type == "Production") %>% 
  left_join(full_qtl10k %>% 
              select(CHR, BP, QTL = trait_ID, QTL_ID)) %>% 
  left_join(full_gene10k %>% 
              select(CHR, BP, gene_id, gene_name)) 
```

```{r, eval = FALSE}
full_gene10k %>% 
  filter(str_detect(gene_name, "GH"))
```

# Chromosome 5

## Lead peak

```{r, eval = FALSE}
fixed9_snp1101 %>% 
  filter(neglog10q > q_fdr1) %>% 
  filter(chr == 5) %>% 
  filter(21000000 > pos) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(full_gene10k) %>% 
  group_by(gene_name) %>% 
  tally(sort = TRUE)
```

### Lead SNP

```{r, eval = FALSE}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  filter(chr == 5) %>% 
  filter(pos == 18767155) %>% 
  select(CHR = chr, BP = pos) %>% 
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = .,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 415000) 
```

## Second peak

```{r, eval = FALSE}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  filter(chr == 5) %>% 
  filter(pos > 21000000) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(full_gene0) 

```

```{r, eval = FALSE}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  filter(chr == 5) %>% 
  filter(pos > 21000000) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(full_gene10k) %>% 
  group_by(gene_id, gene_name) %>% 
  tally(sort = TRUE)

```

# Chromosome 23

```{r, eval = FALSE}
all_gene10k %>% 
  filter(CHR == 23) %>% 
  select(dataset, everything(), -strand, -width) %>% 
  filter(gene_name == "PRL") %>% 
  filter(BP>= start_pos & end_pos >= BP)
```

```{r, eval = FALSE}
all_gene50k %>% 
  filter(CHR == 23) %>% 
  select(dataset, gene_name, everything(), -strand, -width) %>% 
  arrange(BP)
```

```{r}
fixed9_snp1101 %>% 
  filter(chr == 23) %>%
  filter(pos > 30000000) %>% 
  filter(50000000 > pos) %>%
  left_join(full_gene10k %>% 
              select(gene_name, chr = CHR, pos = BP, start_pos, end_pos)) %>% 
  mutate(clr = case_when(pos >= start_pos & end_pos >= pos & gene_name == "CDKAL1" ~ "black",
                         TRUE ~ "#b9aa97")) %>% 
  ggplot(aes(x = pos,
             y = neglog10q,
             color = clr)) +
  geom_point(alpha = 0.75) +
  scale_color_identity() +
  scale_x_continuous(breaks = c(30000000, 35000000, 40000000, 45000000, 50000000),
                     labels = c("30 Mb", "35 Mb", "40 Mb", "45 Mb", "50 Mb")) +
  theme_classic() +
  theme(axis.title = element_text(size = 18),
        axis.title.y = element_text(margin = margin(t = 0,
                                                    r = 13,
                                                    b = 0,
                                                    l = 0)),
        axis.text = element_text(size = 14),
        legend.text = element_text(size = 14),
        axis.title.x = element_blank(),
        panel.background = element_rect(fill = "transparent",
                                        colour = NA),
        plot.background = element_rect(fill = "transparent",
                                       colour = NA)) +
  labs(x = NULL,
       y = latex2exp::TeX("$-log_{10}(q-value)$"),
       title = NULL) +
  geom_hline(yintercept = 1,
             color = "red",
             size = 0.5)
```

```{r}
ggsave(filename = here::here("figures/annotation/full_q_chr23.cdkal1.transparent.png"),
       width = 5,
       height = 3,
       bg = "transparent")
```

# Supplementary file

```{r, eval = FALSE}
supp1 <-
  all_gene10k %>% 
  select(-CHR) %>% 
  rename(snp_pos = BP, gene_start = start_pos, gene_end = end_pos) %>% 
  left_join(fixed9_snp1101 %>% 
              mutate(dataset = "full",
                     chr = as.character(chr)) %>% 
              filter(neglog10q > 1) %>% 
              select(dataset, snp_pos = pos, chr, snp_p = p_value) %>% 
              bind_rows(breeds_snp1101 %>% 
                          mutate(chr = as.character(chr)) %>% 
                          filter(neglog10p > 5) %>% 
                          select(dataset, snp_pos = pos, chr, snp_p = p_value))) %>% 
  mutate(within_gene = case_when(snp_pos >= gene_start & gene_end >= snp_pos ~ TRUE,
                                 TRUE ~ FALSE)) %>% 
  select(dataset, chr, snp_pos, snp_p, gene_start, gene_end, gene_id, gene_name, gene_biotype, within_gene)
```


```{r, eval = FALSE}
write_csv(supp1,
          here::here("talks_articles/s1.genes10kb.csv"),
          na = "")
```

