---
title: "GWAS results annotation & enrichment"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(glue)
library(magrittr)
library(stringr)
library(tidylog)
library(GALLO)
library(gprofiler2)

options(scipen = 999)
```

# Notes & questions

* SNP1101 `freq` column is allele frequency not genotype frequency: $\frac{g_2+\frac{1}{2}g_1}{g_0+g_1+g_2}$ or $\frac{2g_{2}+g_{1}}{2(g_0+g_1+g_2)}$
* GCTA COJO requirements:
    + "`A1` needs to be the effect allele with `A2` being the other allele and `freq` should be the frequency of `A1`
    + `A1` and `A2` are A/T/C/G base calls

# Setup

## SNP1101 raw GWAS results

```{r, warning = FALSE, message=FALSE}
fixed9_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(neglog10p = -log10(p_value),
         abs_snp = abs(b_value),
         neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
  left_join(read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs.txt"),
                        skip = 12) %>%
              select(chr = Chr, pos = Pos, fdr = FDR_GW, contains("g")))
  
```

```{r}
breeds_snp1101 <-
  c("hfd1", "an1", "bg1", "igs1") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue("data/derived_data/snp1101/{.x}/out/gwas_ssr_{.x}_bvs_p.txt")))  %>%
                   janitor::clean_names() %>%
                   mutate(neglog10p = -log10(p_value),
                          neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
                   left_join(read_table2(here::here(glue("data/derived_data/snp1101/{.x}/out/gwas_ssr_{.x}_bvs.txt")),
                                         skip = 12) %>%
                               select(chr = Chr, pos = Pos, fdr = FDR_GW, a)),
                 .id = "dataset")
```

```{r}
full_pass <-
  fixed9_snp1101 %>% 
  filter(neglog10q >= 1) %>% 
  select(CHR = chr, BP = pos)
```

```{r}
breeds_pass <-
  breeds_snp1101 %>% 
  filter(neglog10p >= 5) %>% 
  select(CHR = chr, BP = pos)
```

## GFF and GTF files

```{r, eval=FALSE}
bim <- 
  read_table2(here::here("data/raw_data/geno_dump/200924_HairShed.850K.qc.bim"),
              col_names = FALSE) %>% 
  select(chr = X1, SNP = X2, pos = X4, A1 = X5, A2 = X6)
```

```{r}
gtf <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.101.gtf"),
                 file_type = "gtf")
```

```{r}
gff <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.QTL.gff"),
                 file_type = "gff")
```

## COJO input

```
srun --account deckerlab -p BioCompute,htc4,hpc5 -t 12:00:00 --mem 100G -c 32 source_functions/gcta_1.93.2beta/gcta64 --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --cojo-file data/derived_data/cojo/fixed9/fixed9.ma --cojo-slct --cojo-gc --out data/derived_data/cojo/fixed9/fixed9 --threads 32
```

```{r, eval=FALSE}
fixed9_snp1101 %>% 
  mutate(N = g0+g1+g2,
         se = sqrt(var/(g0+g1+g2))) %>%
left_join(bim) %>% 
  select(SNP, A1, A2, freq, b = b_value, p = p_value, N) %>% 
  write_delim(here::here("data/derived_data/cojo/fixed9/fixed9.ma"),
              delim = " ")

```

### Hereford

```{r}
read_table2(here::here("data/derived_data/grm_inbreeding/mizzou_hairshed.grm.id"),
            col_names = c("full_reg", "iid")) %>% 
  left_join(read_rds(here::here("data/derived_data/breed_key/breed_key.rds"))) %>% 
  mutate(keep = case_when(cross == "HFD" ~ TRUE,
                          str_detect(full_reg, "HER") ~ TRUE)) %>% 
  filter(keep == TRUE) %>% 
  select(full_reg, iid) %>% 
  write_delim(here::here("data/derived_data/cojo/hfd1/keep_list.hfd1.txt"),
              delim = " ",
              col_names = FALSE)
```


# Annotation

## Genes

```{r}
full_gene10k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

```{r}
full_gene0 <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 0) 
```

### Angus

### IGS

### Hereford

### Brangus

## QTLs

```{r}
qtl10k <-
  find_genes_qtls_around_markers(db_file = gff,
                                 marker_file = pass,
                                 method = c("qtl"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

# Enrichment

## Genes

```{r}
gene_enrich10k <-
  gene10k %>% 
  mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
  distinct(id) %>% 
  pull(id) %>% 
  as.list() %>% 
  purrr::set_names() %>% 
  gost(organism = "btaurus",
       significant = TRUE)
```

```{r}
gene_enrich10k$result %>% 
  arrange(p_value)
```

```{r,eval=FALSE}
gene_enrich10k$result %>% 
  select(-parents) %>% 
  as_tibble() %>% 
  write_csv(path = here::here("data/derived_data/annotation/gene_enrich10k.csv"))
```

```{r}
gostplot(gene_enrich10k)
```

```{r}
publish_gosttable(gene_enrich10k)
```

## QTLs

```{r}

qtl_enrich10k <-
  qtl_enrich(qtl_db = gff,
             qtl_file = qtl10k,
             qtl_type = "Name",
             enrich_type = "genome",
             padj = "BH")
```

```{r}
qtl_enrich10k %>% 
  filter(0.05 >= adj.pval) %>% 
  arrange(adj.pval) %>% 
  select(adj.pval, everything()) 
```

srun --account deckerlab -p BioCompute,htc4,hpc5 -t 48:00:00 --mem 100G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --cojo-file data/derived_data/cojo/fixed9/fixed9.ma --cojo-slct --out data/derived_data/cojo/fixed9/fixed9 --thread-num 56