---
title: "GWAS results annotation & enrichment"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(TissueEnrich)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(glue)
library(magrittr)
library(stringr)
library(tidylog)
library(GALLO)
library(gprofiler2)


options(scipen = 0)
```

# Notes & questions

* SNP1101 `freq` column is allele frequency not genotype frequency: $\frac{g_2+\frac{1}{2}g_1}{g_0+g_1+g_2}$ or $\frac{2g_{2}+g_{1}}{2(g_0+g_1+g_2)}$
* GCTA COJO requirements:
    + "`A1` needs to be the effect allele with `A2` being the other allele and `freq` should be the frequency of `A1`
    + `A1` and `A2` are A/T/C/G base calls
* COJO runs very slowly with `--cojo-gc` option + it calculated a very small inflation factor, took it out

# Setup

## SNP1101 raw GWAS results

```{r, warning = FALSE, message=FALSE}
fixed9_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(neglog10p = -log10(p_value),
         abs_snp = abs(b_value),
         neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
  left_join(read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs.txt"),
                        skip = 12) %>%
              select(chr = Chr, pos = Pos, fdr = FDR_GW, contains("g")))
  
```

```{r}
breeds_snp1101 <-
  c("hfd1", "an1", "bg1", "igs1") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue("data/derived_data/snp1101/{.x}/out/gwas_ssr_{.x}_bvs_p.txt")))  %>%
                   janitor::clean_names() %>%
                   mutate(neglog10p = -log10(p_value),
                          neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
                   left_join(read_table2(here::here(glue("data/derived_data/snp1101/{.x}/out/gwas_ssr_{.x}_bvs.txt")),
                                         skip = 12) %>%
                               select(chr = Chr, pos = Pos, fdr = FDR_GW, contains("g"))),
                 .id = "dataset")
```

```{r}
full_pass <-
  fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  select(CHR = chr, BP = pos)
```

```{r}
breeds_pass <-
  breeds_snp1101 %>% 
  filter(neglog10p > 5) %>% 
  select(dataset, CHR = chr, BP = pos)
```

## BIM, GFF, and GTF files

```{r, eval=FALSE}
bim <- 
  read_table2(here::here("data/raw_data/geno_dump/200924_HairShed.850K.qc.bim"),
              col_names = FALSE) %>% 
  select(chr = X1, SNP = X2, pos = X4, A1 = X5, A2 = X6)
```

```{r}
gtf <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.101.gtf"),
                 file_type = "gtf")
```

```{r}
gff <-
  import_gff_gtf(db_file = here::here("data/raw_data/Bos_taurus.ARS-UCD1.2.QTL.gff"),
                 file_type = "gff")
```

## COJO input

```{r, eval=FALSE}
fixed9_snp1101 %>% 
  mutate(N = g0+g1+g2) %>%
  left_join(bim) %>% 
  select(SNP, A1, A2, freq, b = b_value, var, p = p_value, N) %>% 
  write_delim(here::here("data/derived_data/cojo/fixed9/fixed9.ma"),
              delim = " ")

```

```
srun --account deckerlab -p BioCompute,htc4,hpc5 -t 48:00:00 --mem 100G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --cojo-file data/derived_data/cojo/fixed9/fixed9.ma --cojo-slct --out data/derived_data/cojo/fixed9/fixed9 --threads 56
```

### Angus

```{r}
read_table2(here::here("data/derived_data/grm_inbreeding/mizzou_hairshed.grm.id"),
            col_names = c("full_reg", "iid")) %>% 
  left_join(read_rds(here::here("data/derived_data/breed_key/breed_key.rds"))) %>% 
  mutate(keep = case_when(cross == "AN" ~ TRUE,
                          str_detect(full_reg, "AAA") ~ TRUE,
                          str_detect(full_reg, "BIR") ~ TRUE)) %>% 
  filter(keep == TRUE) %>% 
  select(full_reg, iid) %>% 
  write_delim(here::here("data/derived_data/cojo/an1/keep_list.an1.txt"),
              delim = " ",
              col_names = FALSE)
```

```
module load plink/plink-1.90b
srun -c 12 --account deckerlab plink --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --double-id --cow --threads 12 --keep data/derived_data/cojo/an1/keep_list.an1.txt --make-bed --out data/derived_data/cojo/an1/cojo.an1
```

```{r, eval=FALSE}
breeds_snp1101 %>% 
  filter(dataset == "an1") %>% 
  mutate(N = g0+g1+g2,
         #se = sqrt(var/(g0+g1+g2))) 
         )%>%
  left_join(bim) %>% 
  select(SNP, A1, A2, freq, b = b_value, var, p = p_value, N) %>% 
  filter(freq != 0) %>% 
  write_delim(here::here("data/derived_data/cojo/an1/an1.ma"),
              delim = " ")

```

```
srun --account deckerlab -p BioCompute,htc4,hpc5 -t 48:00:00 --mem 100G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/derived_data/cojo/an1/cojo.an1 --cojo-file data/derived_data/cojo/an1/an1.ma --cojo-slct --out data/derived_data/cojo/an1/an1 --threads 56
```

### Brangus

```{r}
read_table2(here::here("data/derived_data/grm_inbreeding/mizzou_hairshed.grm.id"),
            col_names = c("full_reg", "iid")) %>% 
  left_join(read_rds(here::here("data/derived_data/breed_key/breed_key.rds"))) %>% 
  mutate(keep = case_when(cross == "BGR" ~ TRUE,
                          str_detect(full_reg, "BGR") ~ TRUE)) %>% 
  filter(keep == TRUE) %>% 
  select(full_reg, iid) %>% 
  write_delim(here::here("data/derived_data/cojo/bg1/keep_list.bg1.txt"),
              delim = " ",
              col_names = FALSE)
```

```
module load plink/plink-1.90b
srun -c 12 --account deckerlab plink --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --double-id --cow --threads 12 --keep data/derived_data/cojo/bg1/keep_list.bg1.txt --make-bed --out data/derived_data/cojo/bg1/cojo.bg1
```

```{r, eval=FALSE}
breeds_snp1101 %>% 
  filter(dataset == "bg1") %>% 
  mutate(N = g0+g1+g2,
         #se = sqrt(var/(g0+g1+g2))) 
         )%>%
  left_join(bim) %>% 
  select(SNP, A1, A2, freq, b = b_value, var, p = p_value, N) %>% 
  filter(freq != 0) %>% 
  write_delim(here::here("data/derived_data/cojo/bg1/bg1.ma"),
              delim = " ")

```

```
srun --account deckerlab -p BioCompute,htc4,hpc5 -t 48:00:00 --mem 100G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/derived_data/cojo/bg1/cojo.bg1 --cojo-file data/derived_data/cojo/bg1/bg1.ma --cojo-slct --out data/derived_data/cojo/bg1/bg1 --threads 56
```

### Hereford

```{r, eval=FALSE}
read_table2(here::here("data/derived_data/grm_inbreeding/mizzou_hairshed.grm.id"),
            col_names = c("full_reg", "iid")) %>% 
  left_join(read_rds(here::here("data/derived_data/breed_key/breed_key.rds"))) %>% 
  mutate(keep = case_when(cross == "HFD" ~ TRUE,
                          str_detect(full_reg, "HER") ~ TRUE)) %>% 
  filter(keep == TRUE) %>% 
  select(full_reg, iid) %>% 
  write_delim(here::here("data/derived_data/cojo/hfd1/keep_list.hfd1.txt"),
              delim = " ",
              col_names = FALSE)
```

```
module load plink/plink-1.90b
srun -c 12 --account deckerlab plink --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --double-id --cow --threads 12 --keep data/derived_data/cojo/hfd1/keep_list.hfd1.txt --make-bed --out data/derived_data/cojo/hfd1/cojo.hfd1
```

```{r, eval=FALSE}
breeds_snp1101 %>% 
  filter(dataset == "hfd1") %>% 
  mutate(N = g0+g1+g2,
         #se = sqrt(var/(g0+g1+g2))) 
         )%>%
  left_join(bim) %>% 
  select(SNP, A1, A2, freq, b = b_value, var, p = p_value, N) %>% 
  filter(freq != 0) %>% 
  write_delim(here::here("data/derived_data/cojo/hfd1/hfd1.ma"),
              delim = " ")

```

```
srun --account deckerlab -p BioCompute,htc4,hpc5 -t 48:00:00 --mem 100G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/derived_data/cojo/hfd1/cojo.hfd1 --cojo-file data/derived_data/cojo/hfd1/hfd1.ma --cojo-slct --out data/derived_data/cojo/hfd1/hfd1 --threads 56
```

### IGS

```{r}
read_table2(here::here("data/derived_data/aireml_varcomp/igs1/pull_list.txt"),
            col_names = c("full_reg")) %>% 
  mutate(iid = full_reg) %>% 
  select(full_reg, iid) %>% 
  write_delim(here::here("data/derived_data/cojo/igs1/keep_list.igs1.txt"),
              delim = " ",
              col_names = FALSE)
```

```
module load plink/plink-1.90b
srun -c 12 --account deckerlab plink --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --double-id --cow --threads 12 --keep data/derived_data/cojo/igs1/keep_list.igs1.txt --make-bed --out data/derived_data/cojo/igs1/cojo.igs1
```

```{r, eval=FALSE}
breeds_snp1101 %>% 
  filter(dataset == "igs1") %>% 
  mutate(N = g0+g1+g2,
         #se = sqrt(var/(g0+g1+g2))) 
         )%>%
  left_join(bim) %>% 
  select(SNP, A1, A2, freq, b = b_value, var, p = p_value, N) %>% 
  filter(freq != 0) %>% 
  write_delim(here::here("data/derived_data/cojo/igs1/igs1.ma"),
              delim = " ")

```

```
srun --account deckerlab -p BioCompute,htc4,hpc5 -t 48:00:00 --mem 100G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/derived_data/cojo/igs1/cojo.igs1 --cojo-file data/derived_data/cojo/igs1/igs1.ma --cojo-slct --out data/derived_data/cojo/igs1/igs1 --threads 56
```

# Annotation

## Genes

```{r}
full_gene50k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = full_pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 50000) 
```

```{r}
full_gene10k <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = full_pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

```{r}
full_gene0 <-
  find_genes_qtls_around_markers(db_file = gtf,
                                 marker_file = full_pass,
                                 method = c("gene"),
                                 marker = c("snp"),
                                 interval = 0) 
```

```{r}
breeds_gene50k <-
  map_dfr(.x = c("an1", "bg1", "hfd1", "igs1"),
          ~ breeds_pass %>% 
            filter(dataset == .x) %>% 
            select(-dataset) %>% 
            find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = .,
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 50000) %>% 
            mutate(dataset = .x))
```

```{r}
breeds_gene10k <-
  map_dfr(.x = c("an1", "bg1", "hfd1", "igs1"),
          ~ breeds_pass %>% 
            filter(dataset == .x) %>% 
            select(-dataset) %>% 
            find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = .,
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 10000) %>% 
            mutate(dataset = .x))
```

```{r}
breeds_gene0 <-
  map_dfr(.x = c("an1", "bg1", "hfd1", "igs1"),
          ~ breeds_pass %>% 
            filter(dataset == .x) %>% 
            select(-dataset) %>% 
            find_genes_qtls_around_markers(db_file = gtf,
                                           marker_file = .,
                                           method = c("gene"),
                                           marker = c("snp"),
                                           interval = 0) %>% 
            mutate(dataset = .x))
```

```{r}
all_gene10k <-
  full_gene10k %>% 
  mutate(dataset = "full") %>% 
  bind_rows(breeds_gene10k)
```

```{r}
all_gene50k <-
  full_gene50k %>% 
  mutate(dataset = "full") %>% 
  bind_rows(breeds_gene50k)
```

```{r}
full_gene10k %>% 
  group_by(chr, gene_id, gene_name) %>% 
  tally(sort = TRUE)
```

## QTLs

```{r}
full_qtl10k <-
  find_genes_qtls_around_markers(db_file = gff,
                                 marker_file = full_pass,
                                 method = c("qtl"),
                                 marker = c("snp"),
                                 interval = 10000) 
```

```{r}
breeds_qtl10k <-
  purrr::map_dfr(.x = c("an1", "bg1", "hfd1", "igs1"),
                 ~ breeds_pass %>% 
                   filter(dataset == .x) %>% 
                   select(-dataset) %>% 
                   find_genes_qtls_around_markers(db_file = gff,
                                                  marker_file = .,
                                                  method = c("qtl"),
                                                  marker = c("snp"),
                                                  interval = 10000) %>% 
                   mutate(dataset = .x))
   
```

```{r}
breeds_gene_enrich10k[[1]]$result %>% 
  View()
```


# Calculate enrichments

## Genes

```{r}

full_gene_enrich10k <-
  full_gene10k %>% 
  mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
  distinct(id) %>% 
  pull(id) %>% 
  as.list() %>% 
  purrr::set_names() %>% 
  gost(organism = "btaurus",
       significant = TRUE)
```

```{r}

breeds_gene_enrich10k <-
  purrr::map(.x = c("an1", "bg1", "hfd1", "igs1"),
             ~ breeds_gene10k %>% 
               filter(dataset == .x) %>% 
               mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
               distinct(id) %>% 
               pull(id) %>% 
               as.list() %>% 
               purrr::set_names() %>% 
               gost(organism = "btaurus",
                    significant = TRUE))
  
```

```{r}
shared_gene_enrich_10k <-
  all_gene10k %>% 
  group_by(gene_id) %>% 
  filter(n_distinct(dataset) > 1) %>% 
  ungroup() %>% 
  mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
  distinct(id) %>% 
  pull(id) %>% 
  as.list() %>% 
  purrr::set_names() %>% 
  gost(organism = "btaurus",
       significant = TRUE)

```

## QTLs

```{r}

full_qtl_enrich10k <-
  qtl_enrich(qtl_db = gff,
             qtl_file = full_qtl10k,
             qtl_type = "Name",
             enrich_type = "genome",
             padj = "BH")
```

```{r}

breeds_qtl_enrich10k <-
  purrr::map_dfr(.x = c("an1", "bg1", "hfd1", "igs1"),
                 ~ breeds_qtl10k %>% 
                   filter(dataset == .x) %>% 
                   select(-dataset) %>% 
                   qtl_enrich(qtl_db = gff,
                              qtl_file = .,
                              qtl_type = "Name",
                              enrich_type = "genome",
                              padj = "BH") %>% 
                   mutate(dataset = .x))
  
```

# Gene ontology enrichment results

```{r}
shared_gene_enrich_10k$result
```


```{r}
gostplot(full_gene_enrich10k)
```

```{r}
publish_gosttable(full_gene_enrich10k)
```

## Testosterone?

```{r}
all_go_results %>% 
  group_by(term_name) %>% 
  filter(str_detect(term_name, "testosterone")) 
  
```

# QTL enrichment results

## Full dataset

```{r}
full_qtl_enrich10k %>% 
  View()
  filter(0.05 >= adj.pval) %>% 
  arrange(adj.pval) %>% 
  select(adj.pval, everything()) %>% 
  View()
```

## Breed-specific datasets

```{r}
breeds_qtl_enrich10k %>% 
  filter(0.05 > adj.pval) %>% 
  select(dataset, adj.pval, everything()) %>% 
  arrange(dataset, adj.pval) %>% 
  #filter(N_QTLs > 1) %>% 
  select(dataset:N_QTLs_db) %>% 
  View()
```

## White spotting?

```{r}
full_qtl10k %>% 
  filter(trait_ID == "White spotting") %>% 
  select(-start_pos, -end_pos) %>% 
  left_join(full_gene0 %>% 
              filter(CHR == 22) %>% 
              select(CHR, BP, gene_name, gene_id, start_pos, end_pos)) %>% 
  filter(between(BP, start_pos, end_pos))
  
  
  
```

```{r}
full_gene0 %>% 
  filter(CHR == 22) %>% 
  filter(between(BP, start_pos, end_pos))
```


```{r}
full_qtl10k %>% 
  filter(trait_ID == "Eye area pigmentation") %>% 
  select(-start_pos, -end_pos) %>% 
  left_join(full_gene0 %>% 
              filter(CHR == 22) %>% 
              select(CHR, BP, gene_name, gene_id, start_pos, end_pos)) 
  
  
  
```

# Chromosome 5

## Lead peak

```{r}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  filter(chr == 5) %>% 
  filter(21000000 > pos) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(full_gene10k) %>% 
  group_by(gene_name) %>% 
  tally(sort = TRUE)
```

## Second peak

```{r}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  filter(chr == 5) %>% 
  filter(pos > 21000000) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(full_gene0) %>% 
  group_by(gene_id, gene_name) %>% 
  tally(sort = TRUE)

```

```{r}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  filter(chr == 5) %>% 
  filter(pos > 21000000) %>% 
  arrange(desc(neglog10q)) %>% 
  select(neglog10q, CHR = chr, BP = pos) %>% 
  left_join(full_gene10k) %>% 
  group_by(gene_id, gene_name) %>% 
  tally(sort = TRUE)

```

# Chromosome 23

```{r}
all_gene10k %>% 
  filter(CHR == 23) %>% 
  select(dataset, everything(), -strand, -width) %>% 
  filter(gene_name == "PRL")
  filter(BP>= start_pos & end_pos >= BP)
```

# Chromosome 8 BG

```{r}
breeds_snp1101 %>% 
  filter(neglog10p > 5) %>% 
  filter(dataset == "bg1") %>% 
  filter(chr == 8) %>% 
  select(neglog10p, CHR = chr, BP = pos) %>% 
  left_join(breeds_gene10k)
```

# Chromosome 9 HFD

```{r}
breeds_snp1101 %>% 
  filter(neglog10p > 5) %>% 
  filter(dataset == "hfd1") %>% 
  filter(chr == 9) %>% 
  select(neglog10p, CHR = chr, BP = pos) %>% 
  left_join(breeds_gene10k)
```

# String DB

```{r}
full_gene10k %>% 
  mutate(id = if_else(!is.na(gene_name), gene_name, gene_id),
         # Try ENSEMBL IDs for genes whose names returned no hits/might be parsing errors
         id = case_when(id %in% c("5S_rRNA", "U6", "TAFA1", "TAFA4", "WFDC1", "SNORA50C", "SNORD104") ~ gene_id,
                        TRUE ~ id)) %>% 
  group_by(id) %>% 
  tally(sort = TRUE) %>% 
  select(id) %>% 
  write_delim(here::here("data/derived_data/annotation/string_genes.txt"),
              col_names = FALSE)
```

```{r}
breeds_gene10k %>% 
  filter(dataset == "an1") %>% 
  mutate(id = if_else(!is.na(gene_name), gene_name, gene_id),
         # Try ENSEMBL IDs for genes whose names returned no hits/might be parsing errors
         id = case_when(id %in% c("5S_rRNA", "U6", "TAFA1", "TAFA4", "WFDC1", "SNORA50C", "SNORD104", "MUCL", "SCRT2") ~ gene_id,
                        TRUE ~ id)) %>% 
  group_by(id) %>% 
  tally(sort = TRUE) %>% 
  select(id) %>% 
  write_delim(here::here("data/derived_data/annotation/string_genes.an1.txt"),
              col_names = FALSE)
```

# `EnrichmentMap`

```{r}
gostres <-
  purrr::map(.x = c("full", "an1", "bg1", "hfd1", "igs1"),
             ~ all_gene10k %>% 
               filter(dataset == .x) %>% 
               mutate(id = if_else(!is.na(gene_name), gene_name, gene_id)) %>%
               distinct(id) %>% 
               pull(id) %>% 
               as.list() %>% 
               purrr::set_names() %>% 
               gost(organism = "btaurus",
                    evcodes = TRUE, 
                    multi_query = FALSE,
                    sources = c("GO", "REAC", "MIRNA", "CORUM", "HP", "HPA", "WP")))

```

```{r}
write_gem <-
  function(x, y) {
               gem <- gostres[[x]]$result[,c("term_id", "term_name", "p_value", "intersection")]
               
               colnames(gem) <- c("GO.ID", "Description", "p.Val", "Genes")
               
               gem$FDR <- gem$p.Val
               
               gem$Phenotype = "+1"
               
               gem <- gem[,c("GO.ID", "Description", "p.Val", "FDR", "Phenotype", "Genes")]
               
               write.table(gem,
                           file = here::here(glue("data/derived_data/annotation/enrichment_map/{y}_gem.txt")),
                           sep = "\t",
                           quote = F,
                           row.names = F)
               }
```

```{r}
purrr::walk2(.x = c(1:5), 
             .y = c("full", "an1", "bg1", "hfd1", "igs1"),
             ~ write_gem(x = .x, y = .y))
```

# `TissueEnrich`

First, convert gene names from cattle to human and mouse using `gorth()`.

```{r}
orthologs <-
  c("hsapiens", "mmusculus") %>% 
  purrr::map_dfr(~ all_gene50k %>% 
                   mutate(id = if_else(!is.na(gene_name),
                                       gene_name,
                                       gene_id)) %>%
                   distinct(id) %>% 
                   pull(id) %>% 
                   gorth(query = .,
                         source_organism = "btaurus",
                         target_organism = .x,
                         mthreshold = 1,
                         filter_na = TRUE) %>% 
                   mutate(species = .x)) %>% 
  group_by(input, species) %>% 
  arrange(ensg_number) %>% 
  slice(1) %>% 
  ungroup()
  
```

```{r}

mm_gs <-
  orthologs %>% 
  filter(species == "mmusculus") %>% 
  distinct(ortholog_ensg) %>% 
  pull(ortholog_ensg) %>% 
  GeneSet(geneIds = .,
          organism = "Mus Musculus",
          geneIdType = ENSEMBLIdentifier())
  
```

```{r}
mm_te <- 
  teEnrichment(inputGenes = mm_gs,
               tissueSpecificGeneType = 3,
               rnaSeqDataset = 3)
```

```{r}

hs_gs <-
  orthologs %>% 
  filter(species == "hsapiens") %>% 
  distinct(ortholog_ensg) %>% 
  pull(ortholog_ensg) %>% 
  GeneSet(geneIds = .,
          organism = "Homo Sapiens",
          geneIdType = ENSEMBLIdentifier())
  
```

```{r}
hs_te_hpa <- 
  teEnrichment(inputGenes = hs_gs,
               tissueSpecificGeneType = 3,
               rnaSeqDataset = 1)
```

```{r}
hs_te_gtex <- 
  teEnrichment(inputGenes = hs_gs,
               tissueSpecificGeneType = 3,
               rnaSeqDataset = 2)
```

```{r}

transform_te <- 
  function(output){
    seEnrichmentOutput <- output[[1]]
  
    enrichmentOutput <- setNames(data.frame(assay(seEnrichmentOutput),
                                            row.names = rowData(seEnrichmentOutput)[,1]),
                                 colData(seEnrichmentOutput)[,1])
    
    enrichmentOutput$Tissue <- row.names(enrichmentOutput)
    
    enrichmentOutput %<>%
      tibble::remove_rownames() %>% 
      select(Tissue, everything()) %>% 
      arrange(desc(Log10PValue))
    
    return(enrichmentOutput)
    }

```

```{r}
transform_te(mm_te) 
```

```{r}
transform_te(hs_te_hpa) 
```

```{r}
transform_te(hs_te_gtex) 
```

# Supplementary file

```{r}
supp1 <-
  all_gene10k %>% 
  select(-CHR) %>% 
  rename(snp_pos = BP, gene_start = start_pos, gene_end = end_pos) %>% 
  left_join(fixed9_snp1101 %>% 
              mutate(dataset = "full",
                     chr = as.character(chr)) %>% 
              filter(neglog10q > 1) %>% 
              select(dataset, snp_pos = pos, chr, snp_p = p_value) %>% 
              bind_rows(breeds_snp1101 %>% 
                          mutate(chr = as.character(chr)) %>% 
                          filter(neglog10p > 5) %>% 
                          select(dataset, snp_pos = pos, chr, snp_p = p_value))) %>% 
  mutate(within_gene = case_when(snp_pos >= gene_start & gene_end >= snp_pos ~ TRUE,
                                 TRUE ~ FALSE)) %>% 
  select(dataset, chr, snp_pos, snp_p, gene_start, gene_end, gene_id, gene_name, gene_biotype, within_gene)
```


```{r}
write_csv(supp1,
          here::here("talks_articles/s1.genes10kb.csv"),
          na = "")
```

