---
title: "COJO using year-by-year GWAS meta-analysis results"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(magrittr)
library(stringr)
library(tidylog)
library(glue)

source(here::here("source_functions/hair_manhattan.R"))
source(here::here("source_functions/ggqq.R"))
source(here::here("source_functions/gcif.R"))

options(scipen = 0)
```

# Notes & questions

# Setup

## Raw GWAS results

### Year-by-year in GCTA

```{r, eval=FALSE}
years_gwas <-
  c(2016:2019) %>%
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue("data/derived_data/gcta_gwas.years/{.x}/{.x}.mlma"))) %>% 
                   mutate(neglog10q = -log10(qvalue::qvalue(p)$qvalues)),
                 .id = "year") %>% 
  janitor::clean_names() %>% 
  mutate(neglog10p = -log10(p)) %>% 
  rename(pos = bp)
```

### Using de-regressed EBVs in SNP1101

```{r, warning = FALSE, message=FALSE}
fixed9_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(adj_p = gcif(df = .,
                      adjust_p = TRUE,
                      p_col = p_value),
         q = qvalue::qvalue(adj_p)$qvalues,
         neglog10q = -log10(q))
  
```

```{r}
# I can't remember numbers and have to re-calculate this every time so
q_fdr5 <- -log10(.05)
q_fdr1 <- -log10(.01)
```

## `METASOFT` input file

```{r, warning=FALSE, message=FALSE, eval=FALSE}
purrr::map(.x = c("2016", "2017", "2018", "2019"),
             ~ years_gwas %>% 
               filter(year == .x) %>% 
               filter(!is.na(b)) %>% 
               mutate(id = glue("{chr}:{pos}"),
                      b = round(b, digits = 7),
                      se = round(se, digits = 7)) %>%
               select(id, !!rlang::sym(glue("b_{.x}")) := b, !!rlang::sym(glue("se_{.x}")) := se)) %>% 
  purrr::reduce(left_join) %>% 
  mutate_if(is.numeric, as.character) %>% 
  write_tsv(here::here("data/derived_data/metasoft/years/metasoft_in.years.var.txt"),
            col_names = FALSE)
```

```{r, eval=FALSE}
purrr::map(.x = c("2016", "2017", "2018", "2019"),
             ~ years_gwas %>% 
               filter(year == .x) %>% 
               filter(!is.na(b)) %>% 
               mutate(id = glue("{chr}:{pos}"),
                      b = round(b, digits = 7),
                      se = round(se, digits = 7)) %>%
               select(id, !!rlang::sym(glue("b_{.x}")) := b, !!rlang::sym(glue("se_{.x}")) := se)) %>% 
  purrr::reduce(left_join) %>% 
  filter_at(vars(starts_with("b_")), any_vars(is.na(.)))
```

```
srun java -jar source_functions/Metasoft/Metasoft.jar -pvalue_table source_functions/Metasoft/HanEskinPvalueTable.txt -input data/derived_data/metasoft/years/metasoft_in.years.txt -output data/derived_data/metasoft/years/metasoft_out.years.txt -mvalue -mvalue_p_thres 0.01 -log data/derived_data/metasoft/years/metasoft.years.log
```

* Re-run with lambda value calculated in previous run to correct p-values:
    + `-lambda_mean 1.661412`
    + `-lambda_hetero 0.110889`

```
srun java -jar source_functions/Metasoft/Metasoft.jar -pvalue_table source_functions/Metasoft/HanEskinPvalueTable.txt -input data/derived_data/metasoft/years/metasoft_in.years.txt -output data/derived_data/metasoft/years/metasoft_out.years_adj.txt -mvalue -mvalue_p_thres 0.01 -lambda_mean 1.661412 -lambda_hetero 0.110889 -log data/derived_data/metasoft/years/metasoft.years.log
```

## `METASOFT` output file

```{r, eval=TRUE}
years_metasoft <- 
  read_table2(here::here("data/derived_data/metasoft/years/metasoft_out.years_adj.txt"),
              col_types = cols(.default = "d", RSID = "c")) %>% 
  janitor::clean_names() %>% 
  select(-pvalue_be, -contains("fe")) %>% 
  left_join(read_table2(here::here("data/derived_data/metasoft/years/metasoft_out.years_adj.txt"),
                        skip = 1,
                        col_types = "c---------------dddddddd", 
                        col_names = c("rsid", "p16", "p17", "p18", "p19", "m16", "m17", "m18", "m19"))) %>% 
  mutate(neglog10q_re = -log10(qvalue::qvalue(pvalue_re)$qvalues))

```

## Calculate allele frequencies and missingness for use in COJO

* List of animals in at least one of the 2016-2019 GWAS

```{r, eval=FALSE}
c(2016:2019) %>% 
  purrr::map_dfr(~ read_table2(here::here(glue("data/derived_data/gcta_gwas.years/{.x}/pheno.txt")),
                               col_names = FALSE) %>% 
                   select(X1, X2)) %>% 
  distinct() %>% 
  write_tsv(here::here("data/derived_data/cojo/years/years.keep.txt"),
            col_names = FALSE)
```

* Calculate allele frequencies and missingness within that list

```
module load plink/plink-1.90b
srun -c 12 plink --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --double-id --cow --threads 12 --keep data/derived_data/cojo/years/years.keep.txt --freq --missing --out data/derived_data/cojo/years/years
```

* Subset animals to only year-by-year GWAS to try using them as LD ref rather than full dataset

```
module load plink/plink-1.90b
srun -c 12 plink --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --double-id --cow --threads 12 --keep data/derived_data/cojo/years/years.keep.txt --freq --missing --make-bed --out data/derived_data/cojo/years/years_subset
```

```{r, eval=FALSE}
read_table2(here::here("data/derived_data/cojo/years/years.lmiss")) %>% 
  janitor::clean_names() %>% 
  distinct(n_geno)
```

## COJO input file

* Attach frequency & missingness data from 2016-2019 to `METASOFT` output

```{r, eval=FALSE}
years_metasoft %>% 
  filter(!is.na(beta_re)) %>% 
  filter(!is.na(std_re)) %>% 
  left_join(years_gwas %>% 
              distinct(chr, pos, snp) %>% 
              mutate(rsid = glue("{chr}:{pos}"))) %>%
  left_join(read_table2(here::here("data/derived_data/cojo/years/years.frq")) %>%
              janitor::clean_names()) %>% 
  mutate(N = 10316) %>% 
  select(snp, a1, a2, freq = maf, b = beta_re, se = std_re, p = pvalue_re, N) %>% 
  assertr::verify(!is.na(freq)) %>% 
  filter(freq != 0) %>% 
  write_delim(here::here("data/derived_data/cojo/years/years.ma"),
              delim = " ")
```

* Run COJO using only year-by-year GWAS animals as LD ref

```
srun --account animalsci -p BioCompute,htc4,hpc5 -t 12:00:00 --mem 48G -c 56 source_functions/gcta_1.93.2beta/gcta64 --bfile data/derived_data/cojo/years/years_subset --cojo-file data/derived_data/cojo/years/years.ma --cojo-slct --out data/derived_data/cojo/years/years_subset --threads 56
```

## COJO output files

```{r}
years_jma <- 
  read_table2(here::here("data/derived_data/cojo/years/years_subset.jma.cojo")) %>% 
  janitor::clean_names() %>% 
  rename(pos = bp)
```

```{r}
years_cma <- 
  read_table2(here::here("data/derived_data/cojo/years/years_subset.cma.cojo")) %>% 
  janitor::clean_names() %>% 
  rename(pos = bp) %>% 
  mutate(neglog10p_c = -log10(p_c),
         neglog10q_c = -log10(qvalue::qvalue(p_c)$qvalues))
```

# Individual year GWAS results

```{r, fig.width=8, fig.height=9, eval=FALSE}

years_gwas %>% 
  filter(0.1 > p) %>% 
  mutate(year = case_when(year == "2016" ~ "2016 (n = 5,156)",
                          year == "2017" ~ "2017 (n = 4,989)",
                          year == "2018" ~ "2018 (n = 6,896)",
                          year == "2019" ~ "2019 (n = 5,292)")) %>% 
  hair_manhattan(y_var = neglog10q,
                 y_lab = "-log10(q)",
                 sigline = 1,
                 color1 = "#b9aa97",
                 color2 = "#7e756d") +
  facet_wrap(~ year, 
             nrow = 4,
             scales = "fixed")

ggsave(filename = here::here("figures/gcta_gwas.years/16-19.q.png"),
       width = 8,
       height = 9)

```

# Meta-analysis GWAS results

```{r, eval = FALSE}
years_metasoft %>% 
  filter(0.1 > pvalue_re) %>% 
  mutate(chr = as.numeric(str_extract(rsid, "[[:digit:]]{1,2}(?=:)")),
         pos = as.numeric(str_extract(rsid, "(?<=[[:digit:]]{1,2}:)[[:digit:]]+"))) %>% 
  hair_manhattan(y_var = neglog10q_re,
                 y_lab = "Metasoft -log10(q)",
                 color1 = "#b9aa97",
                 color2 = "#7e756d",
                 sigline = q_fdr5)
```

```{r}
years_metasoft %>% 
  gcif(p_col = pvalue_re)
```


# COJO selected SNPs

## Whole-genome

```{r, fig.width=8, fig.height=3}
slct_wg <-
  fixed9_snp1101 %>% 
  filter(0.9 > p_value) %>% 
  left_join(years_jma %>% 
              select(chr, pos) %>% 
              mutate(clr = TRUE)) %>% 
  hair_manhattan(y_var = neglog10q,
                 y_lab = latex2exp::TeX("$-log_{10}(q-value)$"),
                 sigline = q_fdr1,
                 color1 = "#b9aa97",
                 color2 = "#7e756d",
                 highlight = TRUE,
                 highlight_col = clr) +
  geom_hline(yintercept = q_fdr5,
             color = "blue",
             size = 0.5)
```

```{r, fig.width=8, fig.height=3}
slct_wg
```

```{r}
years_cma %>% 
  filter(0.5 > p_c) %>% 
  hair_manhattan(y_var = neglog10q, 
                 y_lab = latex2exp::TeX("$-log_{10}(q-value)$"),
                 sigline = q_fdr1,
                 color1 = "#b9aa97",
                 color2 = "#7e756d") +
  geom_hline(yintercept = q_fdr5, color = "blue")
```


## Chromosome 5

```{r, fig.width=8, fig.height=3}
slct_5 <-
  fixed9_snp1101 %>% 
  filter(chr == 5) %>%
  filter(30000000 >= pos) %>% 
  filter(pos >= 10000000) %>% 
  mutate(clr = "#b9aa97") %>% 
  ggplot(aes(x = pos,
             y = neglog10q,
             color = clr)) +
  geom_point(alpha = 0.75) +
  scale_color_identity() +
  geom_point(data = years_jma %>% 
               filter(chr == 5) %>% 
               select(chr, pos) %>% 
               left_join(fixed9_snp1101 %>% 
                           select(chr, pos, neglog10q)),
             aes(x = pos,
                 y = neglog10q),
             color = "black",
             fill = "black") +
  scale_x_continuous(breaks = c(10000000, 15000000, 20000000, 25000000, 30000000),
                     labels = c("10 Mb", "15 Mb", "20 Mb", "25 Mb", "30 Mb")) +
  theme_classic() +
  theme(axis.title = element_text(size = 18),
        axis.title.y = element_text(margin = margin(t = 0,
                                                    r = 13,
                                                    b = 0,
                                                    l = 0)),
        axis.text = element_text(size = 14),
        legend.text = element_text(size = 14),
        axis.title.x = element_blank(),
        panel.background = element_rect(fill = "transparent",
                                        colour = NA),
        plot.background = element_rect(fill = "transparent",
                                       colour = NA),
        plot.title = element_text(size = 24,
                                      face = "italic",
                                      margin = margin(t = 0,
                                                      r = 0,
                                                      b = 13,
                                                      l = 0))) +
  labs(x = NULL,
       y = latex2exp::TeX("$-log_{10}(q-value)$"),
       title = NULL) +
  geom_hline(yintercept = q_fdr1,
             color = "red",
             size = 0.5) +
  geom_hline(yintercept = q_fdr5,
             color = "blue",
             size = 0.5)
```

```{r, fig.width=8, fig.height=3}
slct_5
```

## Chromosome 23

```{r, fig.width=8, fig.height=3}
slct_23 <-
  fixed9_snp1101 %>% 
  filter(chr == 23) %>%
  filter(pos > 20000000) %>% 
  filter(50000000 > pos) %>%
  mutate(clr = "#b9aa97") %>% 
  ggplot(aes(x = pos,
             y = neglog10q,
             color = clr)) +
  geom_point(alpha = 0.75) +
  scale_color_identity() +
  geom_point(data = years_jma %>% 
               filter(chr == 23) %>% 
               select(chr, pos) %>% 
               left_join(fixed9_snp1101 %>% 
                           select(chr, pos, neglog10q)),
             aes(x = pos,
                 y = neglog10q),
             color = "black",
             fill = "black") +
  scale_x_continuous(breaks = c(20000000, 25000000, 30000000, 35000000, 40000000, 45000000, 50000000),
                     labels = c("20 Mb", "25 Mb", "30 Mb", "35 Mb", "40 Mb", "45 Mb", "50 Mb")) +
  theme_classic() +
  theme(axis.title = element_text(size = 18),
        axis.title.y = element_text(margin = margin(t = 0,
                                                    r = 13,
                                                    b = 0,
                                                    l = 0)),
        axis.text = element_text(size = 14),
        legend.text = element_text(size = 14),
        axis.title.x = element_blank(),
        panel.background = element_rect(fill = "transparent",
                                        colour = NA),
        plot.background = element_rect(fill = "transparent",
                                       colour = NA),
        plot.title = element_text(size = 24,
                                      face = "italic",
                                      margin = margin(t = 0,
                                                      r = 0,
                                                      b = 13,
                                                      l = 0))) +
  labs(x = NULL,
       y = latex2exp::TeX("$-log_{10}(q-value)$"),
       title = NULL) +
  geom_hline(yintercept = q_fdr1,
             color = "red",
             size = 0.5) +
  geom_hline(yintercept = q_fdr5,
             color = "blue",
             size = 0.5)
```

```{r, fig.width=8, fig.height=3}
slct_23
```

## Panel figure

```{r, fig.width=8, fig.height=9, eval = FALSE}


slct_wg/slct_5/slct_23 + plot_annotation(tag_levels = c("a")) & 
  theme(plot.tag = element_text(size = 24),
        plot.margin = margin(t = 0, b = 0, l = 0.75, r = 1, unit = "mm"))

ggsave(filename = here::here("figures/cojo/fixed9snp1101.q.cojo_panel.png"), height = 9, width = 8) 

```