---
title: "General model GWAS using SNP1101"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(patchwork)
library(magrittr)
library(stringr)
library(tidylog)

source(here::here("source_functions/hair_manhattan.R"))
source(here::here("source_functions/ggqq.R"))
source(here::here("source_functions/parse_loglik.R"))
source(here::here("source_functions/lrt_calc.R"))
```

# Notes & questions

* Column names in *_bvs.txt file: SNPID, Chr, Pos, freq, g0, g1, g2, a, b, Var, F, p-value, FDR_GW
* Column names in *_bvs_p.txt file: Chr, Pos, freq, b-value, Var, p-value
* `b` and `b-value` are the same, assuming SNP effect? 
* Guessing g0, g1, g2 are number of individuals with genotypes 0, 1, 2
* What is F?
* What is a?

# Setup

## SNP1101 

```{r, warning = FALSE, message=FALSE}
fixed9_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(neglog10p = -log10(p_value),
         abs_snp = abs(b_value),
         neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
  left_join(read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs.txt"),
                        skip = 12) %>%
              select(chr = Chr, pos = Pos, fdr = FDR_GW, a))
  
```

```{r, warning = FALSE, message=FALSE}
fixed15_snp1101 <-
  read_table2(here::here("data/derived_data/snp1101/fixed15/out/gwas_ssr_fixed15_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(neglog10p = -log10(p_value),
         abs_snp = abs(b_value),
         neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
  left_join(read_table2(here::here("data/derived_data/snp1101/fixed15/out/gwas_ssr_fixed15_bvs.txt"),
                        skip = 12) %>%
              select(chr = Chr, pos = Pos, fdr = FDR_GW, a))
  
```

```{r}
max(fixed9_snp1101$neglog10q)

min(fixed9_snp1101$neglog10q)
```

## GCTA
 
```{r, eval=FALSE}
fixed9_gcta <- 
  read_table2(here::here("data/derived_data/gcta_gwas/fixed9/fixed9.mlma")) %>% 
  janitor::clean_names() %>% 
  mutate(neglog10p = -log10(p),
         neglog10q = -log10(qvalue::qvalue(p)$qvalues)) %>% 
  rename(pos = bp)
```

```{r, eval = FALSE}
fixed15_gcta <-
  read_table2(here::here("data/derived_data/gcta_gwas/fixed15/fixed15.mlma")) %>% 
  janitor::clean_names() %>% 
  mutate(neglog10p = -log10(p),
         neglog10q = -log10(qvalue::qvalue(p)$qvalues)) %>% 
  rename(pos = bp)
```

# Does including first two PCs as fixed effect in the ssGBLUP model provide a better fit?

* Issue: only included phenotypes for genotyped animals in `fixed15` model (can't have missing covariates in BLUPF90), so it isn't really an equal comparison to `fixed9` model. Compare instead to `fixed14` model, which only includes phenotypes for genotyped animals

```{r}
lrt_calc(ll_null = parse_loglik(path = here::here("data/derived_data/aireml_varcomp/fixed14/airemlf90.fixed14.log"),
                                option = "logL"),
         ll_test = parse_loglik(path = here::here("data/derived_data/aireml_varcomp/fixed15/airemlf90.fixed15.log"),
                                option = "logL"))
```

# Summary/exploratory

```{r}
fixed9_snp1101 %>% 
  filter(neglog10q > 1) %>% 
  group_by(chr) %>% 
  summarise(n = n(),
            pct = n()/557) %>% 
  arrange(desc(pct))
```

# Genome-wide Manhattan and Q-Q plots

## Without PC correction

### SNP1101

```{r}
fixed9_snp1101 %>% 
  filter(neglog10q > 1)
```


```{r, fig.width = 11, fig.height = 6.16}

hair_manhattan(df = fixed9_snp1101 %>% 
                 filter(0.5 > p_value),
               y_var = neglog10q, 
               y_lab = latex2exp::TeX("$-log_{10}(q-value)$"),
               sigline = 1,
               color1 = "#b9aa97",
               color2 = "#7e756d")


ggsave(filename = here::here("figures/general_gwas/fixed9_snp1101.q.png"), height = 6.16, width = 11)
```

```{r, fig.width = 11, fig.height = 6.16}

hair_manhattan(df = fixed9_snp1101 %>% 
                 filter(0.01 > p_value),
               y_var = neglog10p, 
               y_lab = latex2exp::TeX("$-log_{10}(p-value)$"),
               sigline = 5,
               color1 = "#b9aa97",
               color2 = "#7e756d")


ggsave(filename = here::here("figures/general_gwas/fixed9_snp1101.p.png"), height = 6.16, width = 11)
```

#### Q-Q plot, all SNPs

```{r}
ggqq(pvector = fixed9_snp1101$p_value)

ggsave(here::here("figures/general_gwas/fixed9_snp1101.Q-Q.png"))
```

#### Q-Q plot, non-significant SNPs

```{r}
ggqq(pvector = fixed9_snp1101[fixed9_snp1101$fdr == -1, "p_value"][[1]])

ggsave(here::here("figures/general_gwas/fixed9_snp1101.Q-Q.nonsig.png"))
```

### GCTA

```{r, fig.width = 11, fig.height = 6.16,eval=FALSE}
hair_manhattan(df = fixed9_gcta %>% 
                 filter(0.01 > p),
               y_var = neglog10q, 
               y_lab = "-log10(q-value)",
               sigline = 1,
               color1 = "#ef8759",
               color2 = "#5da19a")

ggsave(here::here("figures/general_gwas/fixed9_gcta.q.png"))
```

#### Q-Q plot, all SNPs

```{r, eval = FALSE}
ggqq(pvector = fixed9_gcta$p) +
  theme_classic()

ggsave(here::here("figures/general_gwas/fixed9_gcta.Q-Q.png"))
```

## With PC correction

### SNP1101

```{r, fig.width = 11, fig.height = 6.16}

hair_manhattan(df = fixed15_snp1101 %>% 
                 filter(0.01 > p_value),
               y_var = neglog10q, 
               y_lab = "-log10(q-value)",
               sigline = 1,
               color1 = "#b9aa97",
               color2 = "#7e756d") +
  ylim(0, 15)


ggsave(filename = here::here("figures/general_gwas/fixed15_snp1101.q.png"), height = 6.16, width = 11)
```

```{r, fig.width = 11, fig.height = 6.16}

hair_manhattan(df = fixed15_snp1101 %>% 
                 filter(0.01 > p_value),
               y_var = neglog10p, 
               y_lab = "-log10(p-value)",
               sigline = 5,
               color1 = "#b9aa97",
               color2 = "#7e756d") +
  ylim(0, 20)


ggsave(filename = here::here("figures/general_gwas/fixed15_snp1101.p.png"), height = 6.16, width = 11)
```

#### Q-Q plot, all SNPs

```{r}
ggqq(pvector = fixed15_snp1101$p_value)

ggsave(here::here("figures/general_gwas/fixed15_snp1101.Q-Q.png"))
```

### GCTA

```{r, fig.width = 11, fig.height = 6.16, eval = FALSE}
hair_manhattan(df = fixed15_gcta %>% 
                 filter(0.01 > p),
               y_var = neglog10q, 
               y_lab = "-log10(q-value)",
               sigline = 1,
               color1 = "#ef8759",
               color2 = "#5da19a")

ggsave(here::here("figures/general_gwas/fixed15_gcta.q.png"))
```

#### Q-Q plot, all SNPs

```{r, eval = FALSE}
ggqq(pvector = fixed15_gcta$p)

ggsave(here::here("figures/general_gwas/fixed15_gcta.Q-Q.png"))
```

# Chromosme 5 zoom-in

```{r}
fixed9_snp1101 %>% 
  filter(chr == 5) %>% 
  summarise(max(pos))
```


```{r}
fixed9_snp1101 %>% 
  filter(chr == 5) %>%
  ggplot(aes(x = pos,
             y = neglog10q)) +
  geom_point(color = "#7e756d", alpha = 0.75) +
  scale_x_continuous(breaks = c(0, 25000000, 50000000, 75000000, 100000000, 120000000),
                     labels = c("0 Kb", "25 Kb", "50 Kb", "75 Kb", "100 Kb", "120 Kb")) +
  theme_classic() +
  theme(axis.title.y = element_text(margin = margin(t = 0,
                                                    r = 1,
                                                    b = 0,
                                                    l = 0)),
        axis.title.x = element_blank()) +
  labs(x = NULL,
       y = latex2exp::TeX("$-log_{10}(q-value)$"),
       title = NULL) +
  geom_hline(yintercept = 5,
             color = "red",
             size = 0.25)
```

```{r, fig.width=7, fig.height=7}
fixed9_snp1101 %>% 
  filter(chr %in% c(5, 23)) %>% 
  mutate(dataset = "full") %>% 
  bind_rows(breeds_snp1101 %>% 
              filter(chr %in% c(5, 23))) %>%
  arrange(dataset, chr) %>% 
  mutate(dataset = case_when(dataset == "full" ~ "Full",
                             dataset == "an1" ~ "AN",
                             dataset == "bg1" ~ "BG",
                             dataset == "hfd1" ~ "HFD",
                             dataset == "igs1" ~ "IGS"),
         dataset = forcats::fct_relevel(as.factor(dataset), "Full", "AN", "BG", "HFD", "IGS"),
         chr = if_else(chr == 5, "Chromosome 5", "Chromosome 23"),
         chr = forcats::fct_inorder(chr)) %>% 
  ggplot(aes(x = pos,
             y = neglog10p,
             color = dataset)) +
  geom_point(alpha = 0.75) +
  scale_color_manual(values = c("Full" = "#7e756d",
                                "AN" = "#4f6980",
                                "BG" = "#bfbb60",
                                "HFD" = "#f47942",
                                "IGS"= "#638b66"),
                     guide = FALSE) +
  scale_x_continuous(breaks = c(0, 25000000, 50000000, 75000000, 100000000, 120000000),
                     labels = c("0 Kb", "25 Kb", "50 Kb", "75 Kb", "100 Kb", "120 Kb")) +
  theme_classic() +
  theme(axis.title.y = element_text(margin = margin(t = 0,
                                                    r = 1,
                                                    b = 0,
                                                    l = 0)),
        axis.title.x = element_blank()) +
  labs(x = NULL,
       y = latex2exp::TeX("$-log_{10}(p-value)$"),
       title = NULL) +
  geom_hline(yintercept = 5,
             color = "red",
             size = 0.25) +
  facet_grid(dataset ~ chr,
             scales = "free_x")

ggsave(here::here("figures/breeds/chr_zoomin.p.png"), width = 7, height = 7)
```

