---
title: "General model GWAS using SNP1101"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(patchwork)
library(magrittr)
library(stringr)
library(tidylog)

source(here::here("source_functions/calculate_acc.R"))
source(here::here("source_functions/hair_manhattan.R"))
source(here::here("source_functions/ggqq.R"))
```

# Notes & questions

* Column names in *_bvs.txt file: SNPID, Chr, Pos, freq, g0, g1, g2, a, b, Var, F, p-value, FDR_GW
* Column names in *_bvs_p.txt file: Chr, Pos, freq, b-value, Var, p-value
* `b` and `b-value` are the same, assuming SNP effect? 
* Guessing g0, g1, g2 are number of individuals with genotypes 0, 1, 2
* What is F?
* What is a?

*_bvs.txt
Chr     Pos     freq    b-value Var     p-value
1       530998  0.4526  -0.052163       0.00134825      0.146002

*_bvs_p.txt
SNPID   Chr     Pos     freq    g0      g1      g2      a       b       Var     F       p-value FDR_GW
1       1       530998  0.4526  946     1479    654     0.0175772       -0.052163       0.00134825      2.1146  0.146002        -1

# Setup

## SNP1101 results

```{r, warning = FALSE, message=FALSE}
snp_sol <-
  read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs_p.txt"))  %>%
  janitor::clean_names() %>%
  mutate(neglog10p = -log10(p_value),
         abs_snp = abs(b_value),
         neglog10q = -log10(qvalue::qvalue(p_value)$qvalues)) %>%
  left_join(read_table2(here::here("data/derived_data/snp1101/fixed9/out/gwas_ssr_fixed9_bvs.txt"),
                        skip = 12) %>%
              select(chr = Chr, pos = Pos, fdr = FDR_GW, a))
  
```

## GCTA results

```
srun -c 24 source_functions/gcta_1.93.2beta/gcta64 --mlma --bfile data/raw_data/geno_dump/200924_HairShed.850K.qc --pheno /storage/hpc/group/UMAG/WORKING/hjdzpd/mizzou_hairshed/data/derived_data/gcta_gwas/fixed9/pheno.txt --grm-gz data/derived_data/grm_inbreeding/mizzou_hairshed --thread-num 24 --out data/derived_data/gcta_gwas/fixed9/fixed9
```
 
```{r, eval = FALSE}

gctatest <-
  wideDRP(Data = trait,
        animalId = "full_reg",
        sireId = "sire_reg",
        damId = "dam_reg",
        animalEBV = "solution",
        sireEBV = "sire_sol",
        damEBV = "dam_sol",
        animalr2 = "acc",
        sirer2 = "sire_acc",
        damr2 = "dam_acc",
        c = 0.1,
        h2 = h2) %>% 
  left_join(read_table2(here::here("data/derived_data/grm_inbreeding/mizzou_hairshed.grm.id"),
                        col_names = c("full_reg", "iid"))) %>% 
  filter(!is.na(iid))


gctatest %>% 
  filter(!is.na(DRP_Trait)) %>% 
  filter(!is.na(DRP_Trait_w)) %>% 
  mutate(pheno = DRP_Trait*DRP_Trait_w) %>% 
  select(full_reg, iid, pheno, solution) %>% 
  filter(!is.na(pheno)) %>% 
  write_delim(here::here("data/derived_data/gcta_gwas/fixed9/pheno.txt"),
              col_names = FALSE)
```

```{r}
gcta_mlma <- 
  read_table2(here::here("data/derived_data/gcta_gwas/fixed9/fixed9.mlma")) %>% 
  janitor::clean_names() %>% 
  mutate(neglog10p = -log10(p),
         neglog10q = -log10(qvalue::qvalue(p)$qvalues)) %>% 
  rename(pos = bp)
```


```{r, eval = FALSE}
snp_sol %>% 
  filter(fdr != -1)
```

# Manhattan plot of $-log_{10}(q-value)$

## SNP1101

```{r, fig.width = 11, fig.height = 6.16}
manall <- 
  hair_manhattan(df = snp_sol,
                 y_var = neglog10q, 
                 y_lab = "neglog10q",
                 sigline = 1,
                 color1 = "#ef8759",
                 color2 = "#5da19a")


manall

#ggsave(filename = here::here("figures/snp1101_p.png"), height = 6.16, width = 11)
```

## GCTA

```{r}

  hair_manhattan(df = gcta_mlma,
                 y_var = neglog10q, 
                 y_lab = "neglog10q",
                 sigline = 1,
                 color1 = "#ef8759",
                 color2 = "#5da19a")

#ggsave(here::here("figures/gcta_q.png"))
```


# QQ plot

## SNP1101

### All SNPs

```{r}
ggqq(pvector = snp_sol$p_value)
```

### Non-significant SNPs

```{r}
ggqq(pvector = snp_sol[snp_sol$fdr == -1, "p_value"][[1]])
```

## GCTA

```{r}
ggqq(pvector = gcta_mlma$p)
```

