---
title: "GEMMA GxE GWAS"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(tidyr)
library(stringr)
library(dplyr)
library(glue)
library(magrittr)
library(lubridate)
library(ggplot2)

source(here::here("source_functions/hair_manhattan.R"))
source(here::here("source_functions/ggqq.R"))
source(here::here("source_functions/gcif.R"))
```

# Notes & questions

* Fit fixed effects directly in GEMMA, dummy coded
    + Column of 1s for mean
    + Calving season, age group, fescue
* Need to have phenotype file and genotype file in same order
    + ~~Phenotype in fam file - make fam manually then use `--keep` to subset genotypes?~~
    + Set animals in GRM but no phenotypes as NA

# Setup

## Raw output from GEMMA GxE GWAS

```{r}

gxe_gwas <-
  purrr::map2_dfr(.x = rep(c("temp", "day_length"), each = 4),
                 .y = rep(c(2016, 2017, 2018, 2019), times = 2),
                 ~ read_table2(here::here(glue("data/derived_data/gxe_gwas/{.x}/{.y}/result.assoc.txt"))) %>% 
                                 rename(pos = ps) %>% 
                                 mutate(var = .x,
                                        year = .y,
                                        adj_p = gcif(., adjust_p = TRUE),
                                        q = qvalue::qvalue(adj_p)$qvalues,
                                        neglog10q = -log10(q)))
```


```{r}
gxe_gwas %>% 
  mutate(neglog10p = -log10(adj_p)) %>% 
  summarise(max(neglog10q),
            max(neglog10p),
            mean(neglog10q))
```

## `METASOFT` input files

```{r, warning=FALSE, message=FALSE, eval=FALSE}
purrr::map(.x = c(2016, 2017, 2018, 2019),
             ~ gxe_gwas %>% 
             filter(var == "day_length") %>% 
             filter(year == .x) %>% 
             filter(!is.na(beta)) %>% 
             mutate(b = round(beta, digits = 7),
                    se = round(se, digits = 7)) %>%
             select(rs, !!rlang::sym(glue("b_{.x}")) := b, !!rlang::sym(glue("se_{.x}")) := se)) %>% 
  purrr::reduce(left_join) %>% 
  mutate_if(is.numeric, as.character) %>% 
  write_tsv(here::here("data/derived_data/metasoft/day_length/metasoft_in.day_length.txt"),
            col_names = FALSE)
```

```
srun java -jar source_functions/Metasoft/Metasoft.jar -pvalue_table source_functions/Metasoft/HanEskinPvalueTable.txt -input data/derived_data/metasoft/day_length/metasoft_in.day_length.txt -output data/derived_data/metasoft/day_length/metasoft_out.day_length.txt -mvalue -mvalue_p_thres 0.01 -log data/derived_data/metasoft/day_length/metasoft.day_length.log
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
purrr::map(.x = c(2016, 2017, 2018, 2019),
           ~ gxe_gwas %>%
             filter(var == "temp") %>% 
             filter(year == .x) %>% 
             filter(!is.na(beta)) %>% 
             mutate(b = round(beta, digits = 7),
                    se = round(se, digits = 7)) %>%
             select(rs, !!rlang::sym(glue("b_{.x}")) := b, !!rlang::sym(glue("se_{.x}")) := se)) %>% 
  purrr::reduce(left_join) %>% 
  mutate_if(is.numeric, as.character) %>% 
  write_tsv(here::here("data/derived_data/metasoft/temp/metasoft_in.temp.txt"),
            col_names = FALSE)
```

```
srun --account animalsci java -jar source_functions/Metasoft/Metasoft.jar -pvalue_table source_functions/Metasoft/HanEskinPvalueTable.txt -input data/derived_data/metasoft/temp/metasoft_in.temp.txt -output data/derived_data/metasoft/temp/metasoft_out.temp.txt -mvalue -mvalue_p_thres 0.01 -log data/derived_data/metasoft/temp/metasoft.temp.log
```

## `METASOFT` output files

```{r, eval=TRUE, warning=FALSE, message=FALSE}
gxe_metasoft <- 
  c("temp", "day_length") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue("data/derived_data/metasoft/{.x}/metasoft_out.{.x}.txt")),
                               col_types = cols(.default = "d", RSID = "c")) %>% 
                   janitor::clean_names() %>% 
                   select(-pvalue_be, -contains("fe")) %>%
                   left_join(read_table2(here::here(glue("data/derived_data/metasoft/{.x}/metasoft_out.{.x}.txt")),
                                         skip = 1,
                                         col_types = "c---------------dddddddd", 
                                         col_names = c("rsid", "p16", "p17", "p18", "p19", "m16", "m17", "m18", "m19"))) %>% 
                   mutate(neglog10q_re2 = -log10(qvalue::qvalue(pvalue_re2)$qvalues),
                          neglog10q_re = -log10(qvalue::qvalue(pvalue_re)$qvalues),
                          chr = as.numeric(str_extract(rsid, "^[[:digit:]]{1,2}(?=:)")),
                          pos = as.numeric(str_extract(rsid, "(?<=[[:digit:]]{1,2}:)[[:digit:]]+(?=:)"))),
                 .id = "var")
  

```

# Diagnostic


## P-value histograms & Q-Q plots prior to genomic control {.tabset}

```{r}
hists <-
  gxe_gwas %>% 
  select(var, year, p_wald) %>% 
  nest(p_wald) %>% 
  mutate(hist = purrr::pmap(list(x = data, y = year, z = var),
                            .f = function(x, y, z) {
                              x %>% 
                                ggplot(aes(x = p_wald)) +
                                geom_histogram() +
                                theme_classic() +
                                labs(title = glue("Wald p-values: {y} {z}"))
                              }),
         qq = purrr::map(.x = data, 
                         ~ ggqq(pvector = .x$p_wald))) %>% 
  select(-data)
```

### 2016 temperature

```{r}
hists %>% 
  filter(year == 2016 & var == "temp") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2016 & var == "temp") %>% 
  pull(qq)
```

### 2016 day length

```{r}
hists %>% 
  filter(year == 2016 & var == "day_length") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2016 & var == "day_length") %>% 
  pull(qq)
```

### 2017 temperature

```{r}
hists %>% 
  filter(year == 2017 & var == "temp") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2017 & var == "temp") %>% 
  pull(qq)
```

### 2017 day length

```{r}
hists %>% 
  filter(year == 2017 & var == "day_length") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2017 & var == "day_length") %>% 
  pull(qq)
```

### 2018 temperature

```{r}
hists %>% 
  filter(year == 2018 & var == "temp") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2018 & var == "temp") %>% 
  pull(qq)
```

### 2018 day length

```{r}
hists %>% 
  filter(year == 2018 & var == "day_length") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2018 & var == "day_length") %>% 
  pull(qq)
```

### 2019 temperature

```{r}
hists %>% 
  filter(year == 2019 & var == "temp") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2019 & var == "temp") %>% 
  pull(qq)
```

### 2019 day length

```{r}
hists %>% 
  filter(year == 2019 & var == "day_length") %>% 
  pull(hist)
```

```{r, cache=TRUE}
hists %>% 
  filter(year == 2019 & var == "day_length") %>% 
  pull(qq)
```

## Inflation factors prior to genomic control

```{r}
gxe_gwas %>% 
  select(var, year, p_wald) %>% 
  group_by(var, year) %>% 
  nest() %>% 
  mutate(`Inflation factor` = purrr::map_dbl(.x = data,
                                              ~ .x %>% 
                                                gcif())) %>% 
  rename(`Variable` = var,
         Year = year) %>% 
  select(-data) %>% 
  arrange(Year)
```

# `METASOFT` Q-Q plots

> Temperature

```{r}
ggqq(pvector = gxe_metasoft %>% 
       filter(var == "temp") %>% 
       pull(pvalue_re))
```

```{r}
gxe_metasoft %>% 
  filter(var == "temp") %>% 
  filter(!is.na(pvalue_re)) %>% 
  select(p_wald = pvalue_re) %>% 
  gcif()
```

> Day length

```{r}
gxe_metasoft %>% 
  filter(var == "day_length") %>% 
  filter(!is.na(pvalue_re)) %>% 
  select(p_wald = pvalue_re) %>% 
  gcif()
```

```{r}
ggqq(pvector = gxe_metasoft %>% 
       filter(var == "day_length") %>% 
       pull(pvalue_re))
```

# `METASOFT` results

```{r}
gxe_metasoft %>% 
  filter(var == "day_length") %>% 
  filter(0.5 > pvalue_re) %>% 
  hair_manhattan(y_var = neglog10q_re,
                 color1 = "#b9aa97",
                 color2 = "#7e756d",
                 sigline = 1)
```

```{r}
gxe_gwas %>% 
  filter(0.1 > adj_p) %>% 
  filter(var == "day_length" & year == "2019") %>% 
  mutate(neglog10p = -log10(adj_p)) %>% 
  hair_manhattan(y_var = neglog10p,
                 color1 = "#b9aa97",
                 color2 = "#7e756d",
                 sigline = 5)
```

```{r}
read_table2(here::here("data/derived_data/aireml_varcomp/fixed9/solutions"),
                         skip = 1,
                         col_names = c("trait", "effect", "id_renamed", "solution", "se")) %>% 
  filter(effect == 1)
```

