---
title: "GEMMA GxE GWAS"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(tidyr)
library(stringr)
library(dplyr)
library(glue)
library(magrittr)
library(lubridate)
library(ggplot2)

source(here::here("source_functions/hair_manhattan.R"))
source(here::here("source_functions/ggqq.R"))
source(here::here("source_functions/gcif.R"))

# I can't remember numbers and have to re-calculate this every time so
q_fdr5 <- -log10(.05)
q_fdr1 <- -log10(.01)
```

# Notes & questions

* ~~Fit fixed effects directly in GEMMA, dummy coded~~ Calculate breeding values using AIREML, subtract estimated CG BLUE from phenotype --> use that adjusted phenotype in GEMMA
    + Column of 1s for mean in GEMMA
    + ~~Calving season, age group, fescue~~
* Okay that I calculated CG effect in one go with all years data rather than 4 separate analyses? Figured it would be more accurate as more information included
* Need to have phenotype file and genotype file in same order
    + ~~Phenotype in fam file - make fam manually then use `--keep` to subset genotypes?~~
    + Set animals in GRM but no phenotypes as NA

---  
    
Tried [GxE test using robust variance estimates](https://epstein-software.github.io/robust-joint-interaction/) before realizing it didn't fit any sort of relatedness matrix. Could fit PCs instead?

* Unlike GEMMA, needs every individual to have a phenotype so subset PLINK files to 2018 day length for testing purposes
* Installed `Rserve` via conda (otherwise, would have to run in an interactive session on Lewis)

```
# Default PLINK module doesn't allow R plugins
module load plink/plink-high-contig-1.90p
R CMD Rserve --no-save
plink --bfile data/derived_data/robust_joint_gxe/test --threads 12 --allow-no-sex --cow --double-id --covar data/derived_data/robust_joint_gxe/test.cov --R source_functions/robust-joint-int-plugin.R
```

# Setup

## GEMMA GxE GWAS

```{r}

gxe_gwas <-
  purrr::map2_dfr(.x = rep(c("temp", "day_length"), each = 4),
                 .y = rep(c(2016, 2017, 2018, 2019), times = 2),
                 ~ read_table2(here::here(glue("data/derived_data/gxe_gwas/{.x}/{.y}/result.assoc.txt"))) %>% 
                                 rename(pos = ps) %>% 
                                 mutate(var = .x,
                                        year = .y,
                                        adj_p = gcif(.,
                                                     adjust_p = TRUE),
                                        q = qvalue::qvalue(adj_p)$qvalues,
                                        neglog10q = -log10(q)))
```

## Metasoft

* Input files

```{r, warning=FALSE, message=FALSE, eval=FALSE}
purrr::map(.x = c(2016, 2017, 2018, 2019),
             ~ gxe_gwas %>% 
             filter(var == "day_length") %>% 
             filter(year == .x) %>% 
             filter(!is.na(beta)) %>% 
             mutate(b = round(beta, digits = 7),
                    se = round(se, digits = 7)) %>%
             select(rs, !!rlang::sym(glue("b_{.x}")) := b, !!rlang::sym(glue("se_{.x}")) := se)) %>% 
  purrr::reduce(left_join) %>% 
  mutate_if(is.numeric, as.character) %>% 
  write_tsv(here::here("data/derived_data/metasoft/day_length/metasoft_in.day_length.txt"),
            col_names = FALSE)
```

```
srun java -jar source_functions/Metasoft/Metasoft.jar -pvalue_table source_functions/Metasoft/HanEskinPvalueTable.txt -input data/derived_data/metasoft/day_length/metasoft_in.day_length.txt -output data/derived_data/metasoft/day_length/metasoft_out.day_length.txt -mvalue -mvalue_p_thres 0.01 -log data/derived_data/metasoft/day_length/metasoft.day_length.log
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
purrr::map(.x = c(2016, 2017, 2018, 2019),
           ~ gxe_gwas %>%
             filter(var == "temp") %>% 
             filter(year == .x) %>% 
             filter(!is.na(beta)) %>% 
             mutate(b = round(beta, digits = 7),
                    se = round(se, digits = 7)) %>%
             select(rs, !!rlang::sym(glue("b_{.x}")) := b, !!rlang::sym(glue("se_{.x}")) := se)) %>% 
  purrr::reduce(left_join) %>% 
  mutate_if(is.numeric, as.character) %>% 
  write_tsv(here::here("data/derived_data/metasoft/temp/metasoft_in.temp.txt"),
            col_names = FALSE)
```

```
srun --account animalsci java -jar source_functions/Metasoft/Metasoft.jar -pvalue_table source_functions/Metasoft/HanEskinPvalueTable.txt -input data/derived_data/metasoft/temp/metasoft_in.temp.txt -output data/derived_data/metasoft/temp/metasoft_out.temp.txt -mvalue -mvalue_p_thres 0.01 -log data/derived_data/metasoft/temp/metasoft.temp.log
```

* Output files

```{r, eval=FALSE, warning=FALSE, message=FALSE}
gxe_metasoft <- 
  c("temp", "day_length") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue("data/derived_data/metasoft/{.x}/metasoft_out.{.x}.txt")),
                               col_types = cols(.default = "d", RSID = "c")) %>% 
                   janitor::clean_names() %>% 
                   select(-pvalue_be, -contains("fe")) %>%
                   left_join(read_table2(here::here(glue("data/derived_data/metasoft/{.x}/metasoft_out.{.x}.txt")),
                                         skip = 1,
                                         col_types = "c---------------dddddddd", 
                                         col_names = c("rsid", "p16", "p17", "p18", "p19", "m16", "m17", "m18", "m19"))) %>% 
                   mutate(neglog10q_re2 = -log10(qvalue::qvalue(pvalue_re2)$qvalues),
                          neglog10q_re = -log10(qvalue::qvalue(pvalue_re)$qvalues),
                          chr = as.numeric(str_extract(rsid, "^[[:digit:]]{1,2}(?=:)")),
                          pos = as.numeric(str_extract(rsid, "(?<=[[:digit:]]{1,2}:)[[:digit:]]+(?=:)"))),
                 .id = "var")
  

```

# Step 1. Calculate contemporary group BLUEs to pre-adjust hair shedding scores

* One single evaluation using all data 2012-2020
* Fit single contemporary group effect: `{farm_id + year + calving_season + age_group + toxic_fescue}` (include `farm_id` but no `score_group`)
    + This ends up meaning that animals with multiple records in the same year have multiple records in a single contemporary group - should I limit to one record per year by randomly choosing a record? Or, add `score_group` back in?

# Step 2. Fit four year-specific GxE GWAS for each day length and temperature using GEMMA

* Using results from in step 1, subtract associated CG BLUE from hair score
* For animals with multiple records in the same year, randomly choose one record

## Diagnostic

### Q-Q plots pre- and post-genomic control {.tabset}

```{r}
qq <-
  gxe_gwas %>% 
  select(var, year, p_wald, adj_p) %>% 
  nest(-var, -year) %>% 
  mutate(pre = purrr::pmap(list(x = data, y = year, z = var),
                            .f = function(x, y, z) {
                              x %>% 
                                pull(p_wald) %>% 
                                ggqq() +
                                labs(title = glue("{y} {z} pre-GC"))
                              }),
         post = purrr::pmap(list(x = data, y = year, z = var),
                            .f = function(x, y, z) {
                              x %>% 
                                pull(adj_p) %>% 
                                ggqq() +
                                labs(title = glue("{y} {z} post-GC"))
                              })) %>% 
  select(-data)
```

#### 2016 temperature

```{r, cache=TRUE}
qq %>% 
  filter(year == 2016 & var == "temp") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2016 & var == "temp") %>% 
  pull(post)
```

#### 2016 day length

```{r, cache=TRUE}
qq %>% 
  filter(year == 2016 & var == "day_length") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2016 & var == "day_length") %>% 
  pull(post)
```

#### 2017 temperature

```{r, cache=TRUE}
qq %>% 
  filter(year == 2017 & var == "temp") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2017 & var == "temp") %>% 
  pull(post)
```

#### 2017 day length

```{r, cache=TRUE}
qq %>% 
  filter(year == 2017 & var == "day_length") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2017 & var == "day_length") %>% 
  pull(post)
```

#### 2018 temperature

```{r, cache=TRUE}
qq %>% 
  filter(year == 2018 & var == "temp") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2018 & var == "temp") %>% 
  pull(post)
```

#### 2018 day length

```{r, cache=TRUE}
qq %>% 
  filter(year == 2018 & var == "day_length") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2018 & var == "day_length") %>% 
  pull(post)
```

#### 2019 temperature

```{r, cache=TRUE}
qq %>% 
  filter(year == 2019 & var == "temp") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2019 & var == "temp") %>% 
  pull(post)
```

#### 2019 day length

```{r, cache=TRUE}
qq %>% 
  filter(year == 2019 & var == "day_length") %>% 
  pull(pre)
```

```{r, cache=TRUE}
qq %>% 
  filter(year == 2019 & var == "day_length") %>% 
  pull(post)
```

### Inflation factors pre- and post-genomic control

```{r}
gxe_gwas %>% 
  select(var, year, p_wald, adj_p) %>% 
  group_by(var, year) %>% 
  nest() %>% 
  mutate(`Pre-GC inflation factor` = purrr::map_dbl(.x = data,
                                              ~ .x %>% 
                                                gcif()),
         `Post-GC inflation factor` = purrr::map_dbl(.x = data,
                                              ~ .x %>% 
                                                gcif(p_col = adj_p))) %>% 
  rename(`Variable` = var,
         Year = year) %>% 
  select(-data) %>% 
  arrange(Year)
```

## Results

```{r, eval=FALSE}
gxe_gwas %>% 
  filter(year == "2016" & var == "day_length") %>% 
  filter(0.1 > adj_p) %>% 
  mutate(neglog10p = -log10(adj_p)) %>% 
  hair_manhattan(y_var = neglog10q, 
                 y_lab = latex2exp::TeX("$-log_{10}(q-value)$"),
                 sigline = q_fdr1,
                 color1 = "#b9aa97",
                 color2 = "#7e756d") +
  geom_hline(yintercept = q_fdr5, color = "blue")
```

# Step 3. Meta-analysis of year-specific GxE GWAS using Metasoft

## Diagnostic 

> Temperature

```{r, eval = FALSE}
ggqq(pvector = gxe_metasoft %>% 
       filter(var == "temp") %>% 
       pull(pvalue_re))
```

```{r, eval = FALSE}
gxe_metasoft %>% 
  filter(var == "temp") %>% 
  filter(!is.na(pvalue_re)) %>% 
  select(p_wald = pvalue_re) %>% 
  gcif()
```

> Day length

```{r, eval = FALSE}
gxe_metasoft %>% 
  filter(var == "day_length") %>% 
  filter(!is.na(pvalue_re)) %>% 
  select(p_wald = pvalue_re) %>% 
  gcif()
```

```{r, eval=FALSE}
ggqq(pvector = gxe_metasoft %>% 
       filter(var == "day_length") %>% 
       pull(pvalue_re))
```

## Results

```{r, eval=FALSE}
gxe_metasoft %>% 
  filter(var == "day_length") %>% 
  filter(0.5 > pvalue_re) %>% 
  hair_manhattan(y_var = neglog10q_re,
                 color1 = "#b9aa97",
                 color2 = "#7e756d",
                 sigline = 1)
```

```{r, eval=FALSE}
gxe_gwas %>% 
  filter(0.1 > adj_p) %>% 
  filter(var == "day_length" & year == "2019") %>% 
  mutate(neglog10p = -log10(adj_p)) %>% 
  hair_manhattan(y_var = neglog10p,
                 color1 = "#b9aa97",
                 color2 = "#7e756d",
                 sigline = 5)
```
